# Beyond Black and White: Adapting models to visual domain shift
### Authors: Chakradhar Rangi

The goal of this project is to demonstrate the need for and application of Domain Adaptation (DA) using the problem of handwritten digit classification. DA techniques allow neural networks trained on a **source domain** to generalize to an unseen **target domain** without access to target labels. We focus on **covariate shift**, where the input distribution changes (grayscale vs. colored/textured) while the classification task remains the same.

In this final notebook, we tackle the limitations of our previous statistical approaches.

## Task 3: Training with Adversarial Domain Adaptation (DANN)

In our previous experiments, we found that distance-based methods (like MMD) improved target accuracy to ~79% but plateaued. While these methods align statistical moments (means/variances), they do not actively hunt for the specific features that distinguish the two domains. To bridge this final gap, we implement **Domain-Adversarial Training of Neural Networks (DANN)**.

Introduced by [Ganin et al. (2016)](https://arxiv.org/abs/1505.07818), this approach draws inspiration from Generative Adversarial Networks (GANs). It introduces a "minimax" game between the feature extractor and a new component: the **Domain Discriminator**. The goal is to train the feature extractor to produce **domain-invariant features**—representations that are discriminative enough to classify digits correctly, yet so generic that the discriminator cannot tell if they came from the Source or Target.

### The Architecture: Three Components
We modify our standard CNN to include three distinct functional blocks:
1.  **Feature Extractor ($G_f$):** Maps inputs to the latent representation $Z$.
2.  **Label Predictor ($G_y$):** The standard classifier that predicts digit labels ($0-9$) from $z$.
3.  **Domain Classifier ($G_d$):** A binary classifier that tries to predict the domain label ($d \in \{0, 1\}$) from $Z$.



### The Gradient Reversal Layer (GRL)
The "adversarial" magic happens via a **Gradient Reversal Layer (GRL)** placed between the Feature Extractor and the Domain Classifier.
* **Forward Pass:** The GRL acts as an identity function (it does nothing).
* **Backward Pass:** The GRL **multiplies the gradient by $-\alpha$**.

This reversal signals the Feature Extractor to minimize the label prediction loss while simultaneously **maximizing the domain classification loss**. In other words, the Feature Extractor learns to "confuse" the Domain Classifier.

### Mathematical Formulation
Let $X_s, Y_s$ be source examples with labels, and $X_t$ be target examples. We assign binary domain labels $d_s=0$ for source data and $d_t=1$ for target data. We optimize two loss functions simultaneously:

**1. Task Loss (Supervised):**
Standard Cross-Entropy loss on the source domain:
$$\mathcal{L}_y(\theta_f, \theta_y) = \mathcal{L}_{\text{CE}}(G_y(G_f(X_s)), Y_s)$$

**2. Domain Loss (Adversarial):**
Binary Cross-Entropy loss on *both* domains:
$$\mathcal{L}_d(\theta_f, \theta_d) = \mathcal{L}_{\text{BCE}}(G_d(G_f(X_s)), d_s) + \mathcal{L}_{\text{BCE}}(G_d(G_f(X_t)), d_t)$$

**The Total Objective:**
$$E(\theta_f, \theta_y, \theta_d) = \mathcal{L}_y(\theta_f, \theta_y) - \alpha \cdot \mathcal{L}_d(\theta_f, \theta_d)$$

*(Note: In our PyTorch implementation, we sum the losses and rely on the GRL to flip the sign of the domain gradient during backpropagation.)*

During training, we search for parameters such that $\theta_d$ minimizes the domain classification error, while $\theta_f$ minimizes the task error but **maximizes** the domain error. The hyperparameter $\alpha$ controls the trade-off, typically annealing from 0 to 1 to allow the feature extractor to stabilize before the adversary becomes too difficult to fool.

In [1]:
# Importing the required libraries
import torch
from torch import nn, optim
from torch.autograd import Function
from torchvision import transforms
from torchvision.datasets import MNIST, ImageFolder
from torch.utils.data import DataLoader, random_split

import numpy as np
import matplotlib.pyplot as plt

import os
import urllib.request
import zipfile
import copy
import time
from tqdm import tqdm

# Check for GPU availability
if torch.cuda.is_available():
    device = torch.device("cuda")
    print(f"Using GPU: {torch.cuda.get_device_name(0)}")
    gpu_name = torch.cuda.get_device_name(0)
    if "A100" in gpu_name:
      # ENABLE TF32
      torch.backends.cuda.matmul.allow_tf32 = True
      torch.backends.cudnn.allow_tf32 = True
else:
    device = torch.device("cpu")
    print("No GPU available, using CPU.")

# Set random seed for reproducibility
def set_seed(seed=42):
    """Sets the seed for reproducibility."""
    np.random.seed(seed)

    # PyTorch
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.backends.cudnn.deterministic = True

    # Apple MPS (Mac)
    if torch.backends.mps.is_available():
        torch.mps.manual_seed(seed)

    print(f"Seed set to {seed}")

set_seed(42)

# Setting global variables
BATCH_SIZE = 256
NUM_EPOCHS = 20
LEARNING_RATE = 1e-3
NUM_CLASSES = 10
NUM_WORKERS = 4

# mount google drive to save and load files
from google.colab import drive
drive.mount('/content/drive')

Using GPU: NVIDIA A100-SXM4-40GB
Seed set to 42


/usr/local/lib/python3.12/dist-packages/torch/backends/__init__.py:46: UserWarning: Please use the new API settings to control TF32 behavior, such as torch.backends.cudnn.conv.fp32_precision = 'tf32' or torch.backends.cuda.matmul.fp32_precision = 'ieee'. Old settings, e.g, torch.backends.cuda.matmul.allow_tf32 = True, torch.backends.cudnn.allow_tf32 = True, allowTF32CuDNN() and allowTF32CuBLAS() will be deprecated after Pytorch 2.9. Please see https://pytorch.org/docs/main/notes/cuda.html#tensorfloat-32-tf32-on-ampere-and-later-devices (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:80.)
  self.setter(val)


Mounted at /content/drive


### 3.1 Let us download and normalize our datasets for the final time!

In [2]:
# Mean and std of MNIST dataset
mnist_mean = (0.1307,) * 3  # Replicated for 3 channels
mnist_std = (0.3081,) * 3  # Replicated for 3 channels

# Define transform
mnist_transform = transforms.Compose([
    transforms.Grayscale(num_output_channels=3),
    transforms.ToTensor(),
    transforms.Normalize(mnist_mean, mnist_std)
])

# Downloading the MNIST dataset from torchvision
mnist_train = MNIST(root='./data', train=True, download=True, transform=mnist_transform)
mnist_test = MNIST(root='./data', train=False, download=True, transform=mnist_transform)

# Split into Train (90%) and Validation (10%)
# Validation is used for model checkpointing/tuning.
train_size = int(0.9 * len(mnist_train))
val_size = len(mnist_train) - train_size
train_subset, val_subset = random_split(mnist_train, [train_size, val_size])

# Define DataLoaders
# pin_memory=True speeds up the transfer of data to the GPU.

source_train_loader = DataLoader(train_subset, batch_size=BATCH_SIZE, shuffle=True, num_workers=NUM_WORKERS, pin_memory=True)
source_val_loader = DataLoader(val_subset, batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS, pin_memory=True)
source_test_loader = DataLoader(mnist_test, batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS, pin_memory=True)

100%|██████████| 9.91M/9.91M [00:00<00:00, 43.5MB/s]
100%|██████████| 28.9k/28.9k [00:00<00:00, 1.08MB/s]
100%|██████████| 1.65M/1.65M [00:00<00:00, 9.16MB/s]
100%|██████████| 4.54k/4.54k [00:00<00:00, 9.75MB/s]


In [3]:
# Downloading MNIST-M dataset from GitHub
# Setup paths
data_root = './data'
zip_name = 'MNIST-M.zip'
zip_path = os.path.join(data_root, zip_name)
url = "https://github.com/mashaan14/MNIST-M/raw/main/MNIST-M.zip"

# Ensure Data Folder Exists
if not os.path.exists(data_root):
    os.makedirs(data_root)
    print(f"Created directory: {data_root}")

# Check if ZIP exists, if not, Download it
if not os.path.exists(zip_path):
    print(f"Downloading {zip_name}...")
    try:
        urllib.request.urlretrieve(url, zip_path)
        print("Download complete.")
    except Exception as e:
        print(f"Download failed: {e}")
else:
    print(f"Found {zip_name}.")

# Force Extraction (to ensure folders exist)
print("Extracting/Verifying data...")
with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    zip_ref.extractall(data_root)

# search for the 'training' folder
found_train_path = None
print(f"\nScanning {data_root} for 'training' folder...")

for root, dirs, files in os.walk(data_root):
    if 'training' in dirs:
        found_train_path = os.path.join(root, 'training')
        print(f"FOUND training folder at: {found_train_path}")
        break

if found_train_path:
    # Set the correct paths dynamically
    train_root = found_train_path
    test_root = found_train_path.replace('training', 'testing')

    print(f"Paths for training and testing:")
    print(f"Train Root: {train_root}")
    print(f"Test Root:  {test_root}")
else:
    print("ERROR: Could not find a 'training' folder after extraction.")
    print("Printing current ./data structure:")
    for root, dirs, files in os.walk(data_root):
        level = root.replace(data_root, '').count(os.sep)
        indent = ' ' * 4 * (level)
        print(f"{indent}{os.path.basename(root)}/")

Download complete.
Extracting/Verifying data...

Scanning ./data for 'training' folder...
FOUND training folder at: ./data/MNIST-M/training
Paths for training and testing:
Train Root: ./data/MNIST-M/training
Test Root:  ./data/MNIST-M/testing


In [4]:
mnistm_mean = (0.4582, 0.4623, 0.4085)
mnistm_std = (0.2386, 0.2239, 0.2444)

mnistm_transform = transforms.Compose([
    transforms.Resize((28, 28)),        # Resize to 28x28 (Original MNIST size)
    transforms.ToTensor(),
    transforms.Normalize(mnistm_mean, mnistm_std)
])

# Load MNIST-M train dataset
mnistm_train = ImageFolder(root=train_root, transform=mnistm_transform)
mnistm_test = ImageFolder(root=test_root, transform=mnistm_transform)

# Define DataLoaders
target_train_loader = DataLoader(mnistm_train, batch_size=BATCH_SIZE, shuffle=True, num_workers=NUM_WORKERS, pin_memory=True)
target_test_loader = DataLoader(mnistm_test, batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS, pin_memory=True)

### 3.2 Let us redefine our model to include adversarial component.

We also have to modify the gradient computation through GRL as discussed before.

In [5]:
# Gradient Reversal Layer
class GradientReversalFn(Function):
    @staticmethod
    def forward(ctx, x, alpha):
        ctx.alpha = alpha
        return x.view_as(x)

    @staticmethod
    def backward(ctx, grad_output):
        # Reverse the gradient and scale by alpha
        output = grad_output.neg() * ctx.alpha
        return output, None

class GradientReversalLayer(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x, alpha=1.0):
        return GradientReversalFn.apply(x, alpha)

In [6]:
class DANN_CNN(nn.Module):
    def __init__(self, num_classes=10):
        super().__init__()

        # Shared Feature Extractor (Same as Task 1/2)
        self.features = nn.Sequential(
            nn.Conv2d(3, 64, kernel_size=3, padding=1),
            nn.BatchNorm2d(64),
            nn.ReLU(inplace=True),
            nn.Conv2d(64, 64, kernel_size=3, padding=1),
            nn.BatchNorm2d(64),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(2),
            nn.Conv2d(64, 128, kernel_size=3, padding=1),
            nn.BatchNorm2d(128),
            nn.ReLU(inplace=True),
            nn.Conv2d(128, 128, kernel_size=3, padding=1),
            nn.BatchNorm2d(128),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(2),
        )

        self.flatten_dim = 128 * 7 * 7

        # Label Predictor
        self.class_classifier = nn.Sequential(
            nn.Linear(self.flatten_dim, 256),
            nn.ReLU(inplace=True),
            nn.Dropout(0.5),
            nn.Linear(256, 128),
            nn.ReLU(inplace=True),
            nn.Linear(128, num_classes)
        )

        # Domain Classifier
        # This branch tries to predict Source (0) vs Target (1)
        self.domain_classifier = nn.Sequential(
            nn.Linear(self.flatten_dim, 100),
            nn.BatchNorm1d(100), # BatchNorm helps adversarial stability
            nn.ReLU(inplace=True),
            nn.Linear(100, 1)    # Binary classification (Src vs Tgt)
        )

        self.grl = GradientReversalLayer()

    def forward(self, x, alpha=1.0):
        # Extract Features
        features = self.features(x)
        features = features.view(features.size(0), -1)

        # Path A: Class Prediction (Standard)
        class_logits = self.class_classifier(features)

        # Path B: Domain Prediction (Adversarial)
        # Apply GRL first! This reverses gradients during backprop.
        features_reversed = self.grl(features, alpha)
        domain_logits = self.domain_classifier(features_reversed)

        return class_logits, domain_logits

def evaluate_accuracy(model, loader, device):
    model.eval()
    correct = 0
    total = 0
    with torch.no_grad():
        for images, labels in loader:
            images, labels = images.to(device), labels.to(device)
            outputs, _ = model(images)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()
    return 100 * correct / total

### 3.3 Training our model: Experiment 1

Now that we have defined the GRL, let us proceed with DA. Let us first pretrain the model with the best CNN model from task 1. We have to be a little careful this time initializing our model as it comes with an additional doman classifier component. We just initialize the digit classifer part of the network with the pretrained model from task 1 and randomly initialize the domain classifer. Additionally, we do not reduce the learning rate in this case as we like the model to also quickly learn domain invariant features.

As for the initialization and training, we shall start with pre-training our model with a batch size of 256 and also fixed $\alpha= 0.25$ unlike annealing as discussed in the beginning.

Here are the summary of parameters for this initial experiment:
1. BATCH SIZE = 256 (initial model) & 256 (training)
2. LEARNING RATE = 1e-3
3. OPTIMIZER = ADAMW
4. LOSS FN = CrossEntropy (class) + BCEWithLogitsLoss (domain)
5. ALPHA = 0.25

In [7]:
LEARNING_RATE = 1e-3
BATCH_SIZE = 256

# Target location of our saved model
models_path = "/content/drive/MyDrive/DA_Project/models/"
best_model_file_path = os.path.join(models_path, f"best_cnn_mnist_{BATCH_SIZE}_{NUM_EPOCHS}_{LEARNING_RATE}.pth")

# Initialize DANN
dann_model = DANN_CNN(num_classes=10).to(device)
if os.path.exists(best_model_file_path):
    print(f"Loading pre-trained source weights from {best_model_file_path}...")
    pretrained_state_dict = torch.load(best_model_file_path, map_location=device)

    # Create a new state_dict for the DANN model, which we'll populate
    dann_state_dict = dann_model.state_dict()

    # Copy weights for the feature extractor
    for k, v in pretrained_state_dict.items():
        if k.startswith('features.'):
            dann_state_dict[k] = v
        elif k.startswith('classifier.'):
            # Rename 'classifier' keys to 'class_classifier' to match DANN_CNN
            new_key = k.replace('classifier.', 'class_classifier.')
            dann_state_dict[new_key] = v

    # Load the modified state_dict. strict=False is used because 'domain_classifier'
    # keys will be missing from the loaded pretrained_state_dict, which is intended.
    dann_model.load_state_dict(dann_state_dict, strict=False)
    print("Loaded pre-trained source weights into DANN_CNN (features and class_classifier). Domain classifier initialized randomly.\n")
else:
    print("Pre-trained weights not found. Starting DANN_CNN from scratch.")

criterion = nn.CrossEntropyLoss()
criterion_domain = nn.BCEWithLogitsLoss()
optimizer = optim.AdamW(dann_model.parameters(), lr=LEARNING_RATE)
best_target_acc = 0.0

print(f" Starting DANN Training (Adversarial)...")
start_time = time.time()

for epoch in range(NUM_EPOCHS):
    dann_model.train()

    # Stats tracking
    total_loss, total_domain_loss, total_class_loss = 0.0, 0.0, 0.0
    total_samples = 0

    # fixed alpha
    alpha = 0.25

    # Zip loaders
    min_len = min(len(source_train_loader), len(target_train_loader))
    loader_zip = tqdm(zip(source_train_loader, target_train_loader), total=min_len)

    for (source_imgs, source_labels), (target_imgs, _) in loader_zip:
        # Setup
        source_imgs, source_labels = source_imgs.to(device), source_labels.to(device)
        target_imgs = target_imgs.to(device)
        batch_size = source_imgs.size(0)    # we do this once again to make sure of the batch size

        # Handle uneven batches
        if source_imgs.size(0) != target_imgs.size(0): continue

        # Create Domain Labels: Source = 0, Target = 1
        domain_label_s = torch.zeros(batch_size,1).float().to(device)
        domain_label_t = torch.ones(batch_size,1).float().to(device)

        optimizer.zero_grad()

        # Forward Pass (Source Data)
        # We calculate BOTH Class Loss and Domain Loss
        class_logits_s, domain_logits_s = dann_model(source_imgs, alpha)
        loss_class = criterion(class_logits_s, source_labels)
        loss_domain_s = criterion_domain(domain_logits_s, domain_label_s)

        # Forward Pass (Target Data)
        # We ONLY calculate Domain Loss (we don't have class labels)
        _, domain_logits_t = dann_model(target_imgs, alpha)
        loss_domain_t = criterion_domain(domain_logits_t, domain_label_t)

        # Backward Pass (Optimization)
        # The GRL inside the model handles the sign flipping automatically!
        loss_domain = (loss_domain_s + loss_domain_t)/2
        loss = loss_class + loss_domain

        loss.backward()
        optimizer.step()

        # Stats
        total_loss += loss.item()
        total_class_loss += loss_class.item()
        total_domain_loss += loss_domain.item()
        total_samples += 1

        with torch.no_grad():
            prob_s = torch.sigmoid(domain_logits_s).mean().item()
            prob_t = torch.sigmoid(domain_logits_t).mean().item()

            # Accuracy checks

            # Source is correctly classified if logit < 0 (Prob < 0.5)
            acc_s = (domain_logits_s < 0).float().mean().item() * 100
            # Target is correctly classified if logit > 0 (Prob > 0.5)
            acc_t = (domain_logits_t > 0).float().mean().item() * 100

            domain_acc = (acc_s + acc_t) / 2

        # Logging
        loader_zip.set_description(
            f"Ep {epoch+1} | Cls: {loss_class.item():.3f} | "
            f"Dom Loss: {loss_domain.item():.3f} | "
            f"Dom Acc: {domain_acc:.1f}%"
        )

        # Average classification loss and domain loss
        avg_class_loss = total_class_loss / total_samples
        avg_domain_loss = total_domain_loss / total_samples

    # Validation
    # Evaluate on Target (Did adaptation work?)
    target_acc = evaluate_accuracy(dann_model, target_test_loader, device)

    # Evaluate on Source (Did we preserve original knowledge?)
    source_acc = evaluate_accuracy(dann_model, source_test_loader, device)

    print(f"Epoch [{epoch+1}/{NUM_EPOCHS}] Stats:")
    print(f"   Avg CL Loss: {avg_class_loss:.4f}")
    print(f"   Avg Domain Loss: {avg_domain_loss:.4f}")
    print(f"   Source Acc: {source_acc:.2f}%")
    print(f"   Target Acc: {target_acc:.2f}%\n")

    # Checkpointing
    if target_acc > best_target_acc:
        best_target_acc = target_acc
        best_model_wts = copy.deepcopy(dann_model.state_dict())

total_time = time.time() - start_time

# Save the best model weights
target_file_path_best = os.path.join(models_path, f"best_DANN_EXP1{BATCH_SIZE}_{NUM_EPOCHS}_{LEARNING_RATE}_{alpha}.pth")
torch.save(best_model_wts, target_file_path_best)
print(f"Best model weights saved to {target_file_path_best}")
print(f"\nAdaptation Complete in {total_time:.1f}s Best Target Accuracy: {best_target_acc:.2f}%")

Loading pre-trained source weights from /content/drive/MyDrive/DA_Project/models/best_cnn_mnist_256_20_0.001.pth...
Loaded pre-trained source weights into DANN_CNN (features and class_classifier). Domain classifier initialized randomly.

 Starting DANN Training (Adversarial)...


Ep 1 | Cls: 0.001 | Dom Loss: 0.170 | Dom Acc: 98.6%: 100%|██████████| 211/211 [00:09<00:00, 21.28it/s]


Epoch [1/20] Stats:
   Avg CL Loss: 0.0189
   Avg Domain Loss: 0.3697
   Source Acc: 99.46%
   Target Acc: 41.53%



Ep 2 | Cls: 0.008 | Dom Loss: 0.110 | Dom Acc: 99.4%: 100%|██████████| 211/211 [00:08<00:00, 25.89it/s]


Epoch [2/20] Stats:
   Avg CL Loss: 0.0228
   Avg Domain Loss: 0.1327
   Source Acc: 99.29%
   Target Acc: 56.28%



Ep 3 | Cls: 0.019 | Dom Loss: 0.080 | Dom Acc: 99.2%: 100%|██████████| 211/211 [00:08<00:00, 25.50it/s]


Epoch [3/20] Stats:
   Avg CL Loss: 0.0192
   Avg Domain Loss: 0.0804
   Source Acc: 99.25%
   Target Acc: 57.81%



Ep 4 | Cls: 0.012 | Dom Loss: 0.058 | Dom Acc: 99.4%: 100%|██████████| 211/211 [00:08<00:00, 26.11it/s]


Epoch [4/20] Stats:
   Avg CL Loss: 0.0185
   Avg Domain Loss: 0.0539
   Source Acc: 99.17%
   Target Acc: 61.73%



Ep 5 | Cls: 0.010 | Dom Loss: 0.071 | Dom Acc: 99.0%: 100%|██████████| 211/211 [00:08<00:00, 26.21it/s]


Epoch [5/20] Stats:
   Avg CL Loss: 0.0199
   Avg Domain Loss: 0.0639
   Source Acc: 99.03%
   Target Acc: 63.72%



Ep 6 | Cls: 0.046 | Dom Loss: 0.244 | Dom Acc: 89.8%: 100%|██████████| 211/211 [00:08<00:00, 25.63it/s]


Epoch [6/20] Stats:
   Avg CL Loss: 0.0278
   Avg Domain Loss: 0.1100
   Source Acc: 97.03%
   Target Acc: 48.13%



Ep 7 | Cls: 0.042 | Dom Loss: 0.095 | Dom Acc: 97.3%: 100%|██████████| 211/211 [00:08<00:00, 25.54it/s]


Epoch [7/20] Stats:
   Avg CL Loss: 0.0423
   Avg Domain Loss: 0.1750
   Source Acc: 98.76%
   Target Acc: 44.34%



Ep 8 | Cls: 0.023 | Dom Loss: 0.043 | Dom Acc: 99.6%: 100%|██████████| 211/211 [00:08<00:00, 25.37it/s]


Epoch [8/20] Stats:
   Avg CL Loss: 0.0322
   Avg Domain Loss: 0.1218
   Source Acc: 99.17%
   Target Acc: 55.44%



Ep 9 | Cls: 0.015 | Dom Loss: 0.076 | Dom Acc: 98.2%: 100%|██████████| 211/211 [00:08<00:00, 25.63it/s]


Epoch [9/20] Stats:
   Avg CL Loss: 0.0188
   Avg Domain Loss: 0.0676
   Source Acc: 99.28%
   Target Acc: 59.37%



Ep 10 | Cls: 0.024 | Dom Loss: 0.111 | Dom Acc: 97.9%: 100%|██████████| 211/211 [00:07<00:00, 26.38it/s]


Epoch [10/20] Stats:
   Avg CL Loss: 0.0279
   Avg Domain Loss: 0.1104
   Source Acc: 98.82%
   Target Acc: 63.40%



Ep 11 | Cls: 0.026 | Dom Loss: 0.171 | Dom Acc: 93.6%: 100%|██████████| 211/211 [00:07<00:00, 26.40it/s]


Epoch [11/20] Stats:
   Avg CL Loss: 0.0364
   Avg Domain Loss: 0.1831
   Source Acc: 99.25%
   Target Acc: 74.47%



Ep 12 | Cls: 0.027 | Dom Loss: 0.431 | Dom Acc: 82.0%: 100%|██████████| 211/211 [00:08<00:00, 26.25it/s]


Epoch [12/20] Stats:
   Avg CL Loss: 0.0456
   Avg Domain Loss: 0.2955
   Source Acc: 98.93%
   Target Acc: 76.62%



Ep 13 | Cls: 0.027 | Dom Loss: 0.510 | Dom Acc: 73.8%: 100%|██████████| 211/211 [00:08<00:00, 25.35it/s]


Epoch [13/20] Stats:
   Avg CL Loss: 0.0501
   Avg Domain Loss: 0.3937
   Source Acc: 99.00%
   Target Acc: 83.50%



Ep 14 | Cls: 0.103 | Dom Loss: 0.498 | Dom Acc: 78.9%: 100%|██████████| 211/211 [00:08<00:00, 25.56it/s]


Epoch [14/20] Stats:
   Avg CL Loss: 0.0590
   Avg Domain Loss: 0.5012
   Source Acc: 98.95%
   Target Acc: 78.69%



Ep 15 | Cls: 0.048 | Dom Loss: 0.585 | Dom Acc: 70.3%: 100%|██████████| 211/211 [00:08<00:00, 25.17it/s]


Epoch [15/20] Stats:
   Avg CL Loss: 0.0526
   Avg Domain Loss: 0.5035
   Source Acc: 98.89%
   Target Acc: 84.66%



Ep 16 | Cls: 0.035 | Dom Loss: 0.409 | Dom Acc: 84.2%: 100%|██████████| 211/211 [00:08<00:00, 25.53it/s]


Epoch [16/20] Stats:
   Avg CL Loss: 0.0495
   Avg Domain Loss: 0.5122
   Source Acc: 99.21%
   Target Acc: 86.78%



Ep 17 | Cls: 0.054 | Dom Loss: 0.608 | Dom Acc: 72.3%: 100%|██████████| 211/211 [00:08<00:00, 25.97it/s]


Epoch [17/20] Stats:
   Avg CL Loss: 0.0436
   Avg Domain Loss: 0.5075
   Source Acc: 98.59%
   Target Acc: 84.29%



Ep 18 | Cls: 0.032 | Dom Loss: 0.805 | Dom Acc: 55.9%: 100%|██████████| 211/211 [00:08<00:00, 25.53it/s]


Epoch [18/20] Stats:
   Avg CL Loss: 0.0452
   Avg Domain Loss: 0.5425
   Source Acc: 99.01%
   Target Acc: 85.38%



Ep 19 | Cls: 0.021 | Dom Loss: 0.489 | Dom Acc: 74.8%: 100%|██████████| 211/211 [00:08<00:00, 25.75it/s]


Epoch [19/20] Stats:
   Avg CL Loss: 0.0445
   Avg Domain Loss: 0.5450
   Source Acc: 99.21%
   Target Acc: 89.13%



Ep 20 | Cls: 0.060 | Dom Loss: 0.546 | Dom Acc: 72.7%: 100%|██████████| 211/211 [00:08<00:00, 25.63it/s]


Epoch [20/20] Stats:
   Avg CL Loss: 0.0421
   Avg Domain Loss: 0.5663
   Source Acc: 99.12%
   Target Acc: 88.20%

Best model weights saved to /content/drive/MyDrive/DA_Project/models/best_DANN_EXP1256_20_0.001_0.25.pth

Adaptation Complete in 214.5s Best Target Accuracy: 89.13%


### Takeaway experiment 1: Optimized DANN

In this primary adversarial experiment, we initializing the DANN with a source model pre-trained on a batch size of 256 and performing adversarial adaptation using the same batch size of 256. We utilized an aggressive learning rate of $1e^{-3}$ and a **fixed weak adversarial strength of $\alpha = 0.25$**.

This configuration yielded the project's state-of-the-art performance, achieving a target accuracy of **89.13%**. This significantly outperforms:
- The Distance-based (MMD) baseline: **79.36%**
- Original DANN paper (2015): **76.1% ± 1.8%**

The superior performance of this training supports two critical findings. First, pretraining the model with a good baseline helps significantly compared to training from scratch (as done in the original DANN paper), as it provides the feature extractor with robust semantic representations before the adversary begins. Second, the fixed weak alpha ($\alpha=0.25$) allowed for "Sufficient Alignment"—aligning semantic digit features while ignoring enough background noise to generalize—without the instability or feature distortion often caused by the stronger adversarial pressure of standard annealing schedules.

### Robustness Check: Re-running Experiment 1 with Multiple Seeds

To confirm that the strong performance of our previous run and was not a fluke, we repeated the exact same setup—pretraining on MNIST with batch size 256 and adapting with batch size 256, fixed $\alpha = 0.25$, and learning rate $1e^{-3}$—across three different random seeds (42, 100, and 2024).

In [8]:
SEEDS = [42, 100, 2024]
LEARNING_RATE = 1e-3
BATCH_SIZE = 256
alpha = 0.25

# Target location of our saved model
models_path = "/content/drive/MyDrive/DA_Project/models/"
best_model_file_path = os.path.join(models_path, f"best_cnn_mnist_{BATCH_SIZE}_{NUM_EPOCHS}_{LEARNING_RATE}.pth")

# Store final results
final_accuracies = []

print(f"   Starting Robustness Experiment on {len(SEEDS)} seeds...")
print(f"   Config: BS={BATCH_SIZE}, LR={LEARNING_RATE}, Alpha={alpha}")

start_time = time.time()

for run_idx, seed in enumerate(SEEDS):
    print(f"\n" + "="*40)
    print(f"  RUN {run_idx+1}/{len(SEEDS)} | SEED: {seed}")
    print("="*40)

    # set seed
    set_seed(seed)

    # Re-Initialize DataLoaders
    source_train_loader = DataLoader(train_subset, batch_size=BATCH_SIZE, shuffle=True, num_workers=NUM_WORKERS, pin_memory=True)
    source_test_loader = DataLoader(mnist_test, batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS, pin_memory=True)
    target_train_loader = DataLoader(mnistm_train, batch_size=BATCH_SIZE, shuffle=True, num_workers=NUM_WORKERS, pin_memory=True)
    target_test_loader = DataLoader(mnistm_test, batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS, pin_memory=True)


    # Initialize DANN
    dann_model = DANN_CNN(num_classes=10).to(device)
    if os.path.exists(best_model_file_path):
        print(f"Loading pre-trained source weights from {best_model_file_path}...")
        pretrained_state_dict = torch.load(best_model_file_path, map_location=device)

        # Create a new state_dict for the DANN model, which we'll populate
        dann_state_dict = dann_model.state_dict()

        # Copy weights for the feature extractor
        for k, v in pretrained_state_dict.items():
            if k.startswith('features.'):
                dann_state_dict[k] = v
            elif k.startswith('classifier.'):
                # Rename 'classifier' keys to 'class_classifier' to match DANN_CNN
                new_key = k.replace('classifier.', 'class_classifier.')
                dann_state_dict[new_key] = v

        # Load the modified state_dict. strict=False is used because 'domain_classifier'
        # keys will be missing from the loaded pretrained_state_dict, which is intended.
        dann_model.load_state_dict(dann_state_dict, strict=False)
        print("Loaded pre-trained source weights into DANN_CNN (features and class_classifier). Domain classifier initialized randomly.\n")
    else:
        print("Pre-trained weights not found. Starting DANN_CNN from scratch.")

    criterion = nn.CrossEntropyLoss()
    criterion_domain = nn.BCEWithLogitsLoss()
    optimizer = optim.AdamW(dann_model.parameters(), lr=LEARNING_RATE)
    best_target_acc = 0.0

    for epoch in range(NUM_EPOCHS):
        dann_model.train()

        # Stats tracking
        total_loss, total_domain_loss, total_class_loss = 0.0, 0.0, 0.0
        total_samples = 0

        # Zip loaders
        min_len = min(len(source_train_loader), len(target_train_loader))
        loader_zip = tqdm(zip(source_train_loader, target_train_loader), total=min_len)

        for (source_imgs, source_labels), (target_imgs, _) in loader_zip:
            # Setup
            source_imgs, source_labels = source_imgs.to(device), source_labels.to(device)
            target_imgs = target_imgs.to(device)
            batch_size = source_imgs.size(0)    # we do this once again to make sure of the batch size

            # Handle uneven batches
            if source_imgs.size(0) != target_imgs.size(0): continue

            # Create Domain Labels: Source = 0, Target = 1
            domain_label_s = torch.zeros(batch_size,1).float().to(device)
            domain_label_t = torch.ones(batch_size,1).float().to(device)

            optimizer.zero_grad()

            # Forward Pass (Source Data)
            # We calculate BOTH Class Loss and Domain Loss
            class_logits_s, domain_logits_s = dann_model(source_imgs, alpha)
            loss_class = criterion(class_logits_s, source_labels)
            loss_domain_s = criterion_domain(domain_logits_s, domain_label_s)

            # Forward Pass (Target Data)
            # We ONLY calculate Domain Loss (we don't have class labels)
            _, domain_logits_t = dann_model(target_imgs, alpha)
            loss_domain_t = criterion_domain(domain_logits_t, domain_label_t)

            # Backward Pass (Optimization)
            # The GRL inside the model handles the sign flipping automatically!
            loss_domain = (loss_domain_s + loss_domain_t)/2
            loss = loss_class + loss_domain

            loss.backward()
            optimizer.step()

            # Stats
            total_loss += loss.item()
            total_class_loss += loss_class.item()
            total_domain_loss += loss_domain.item()
            total_samples += 1

            with torch.no_grad():
                prob_s = torch.sigmoid(domain_logits_s).mean().item()
                prob_t = torch.sigmoid(domain_logits_t).mean().item()

                # Accuracy checks

                # Source is correctly classified if logit < 0 (Prob < 0.5)
                acc_s = (domain_logits_s < 0).float().mean().item() * 100
                # Target is correctly classified if logit > 0 (Prob > 0.5)
                acc_t = (domain_logits_t > 0).float().mean().item() * 100

                domain_acc = (acc_s + acc_t) / 2

            # Logging
            loader_zip.set_description(
                f"Ep {epoch+1} | Cls: {loss_class.item():.3f} | "
                f"Dom Loss: {loss_domain.item():.3f} | "
                f"Dom Acc: {domain_acc:.1f}%"
            )

            # Average classification loss and domain loss
            avg_class_loss = total_class_loss / total_samples
            avg_domain_loss = total_domain_loss / total_samples

        # Validation
        # Evaluate on Target (Did adaptation work?)
        target_acc = evaluate_accuracy(dann_model, target_test_loader, device)

        # Evaluate on Source (Did we preserve original knowledge?)
        source_acc = evaluate_accuracy(dann_model, source_test_loader, device)

        print(f"Epoch [{epoch+1}/{NUM_EPOCHS}] Stats:")
        print(f"   Avg CL Loss: {avg_class_loss:.4f}")
        print(f"   Avg Domain Loss: {avg_domain_loss:.4f}")
        print(f"   Source Acc: {source_acc:.2f}%")
        print(f"   Target Acc: {target_acc:.2f}%\n")

        # Checkpointing
        if target_acc > best_target_acc:
            best_target_acc = target_acc
            best_model_wts = copy.deepcopy(dann_model.state_dict())

    print(f"Run {run_idx+1} Complete. Best Acc: {best_target_acc:.2f}%")
    final_accuracies.append(best_target_acc)

total_time = time.time() - start_time
mean_acc = np.mean(final_accuracies)
std_acc = np.std(final_accuracies)
print(f"\n ROBUSTNESS EXPERIMENT COMPLETE")
print(f"   Individual Runs: {final_accuracies}")
print(f"   Final Result: {mean_acc:.2f}% ± {std_acc:.2f}%")

   Starting Robustness Experiment on 3 seeds...
   Config: BS=256, LR=0.001, Alpha=0.25

  RUN 1/3 | SEED: 42
Seed set to 42
Loading pre-trained source weights from /content/drive/MyDrive/DA_Project/models/best_cnn_mnist_256_20_0.001.pth...
Loaded pre-trained source weights into DANN_CNN (features and class_classifier). Domain classifier initialized randomly.



Ep 1 | Cls: 0.005 | Dom Loss: 0.194 | Dom Acc: 98.2%: 100%|██████████| 211/211 [00:08<00:00, 25.74it/s]


Epoch [1/20] Stats:
   Avg CL Loss: 0.0207
   Avg Domain Loss: 0.3631
   Source Acc: 99.11%
   Target Acc: 41.38%



Ep 2 | Cls: 0.045 | Dom Loss: 0.092 | Dom Acc: 99.6%: 100%|██████████| 211/211 [00:08<00:00, 25.81it/s]


Epoch [2/20] Stats:
   Avg CL Loss: 0.0203
   Avg Domain Loss: 0.1288
   Source Acc: 99.32%
   Target Acc: 47.27%



Ep 3 | Cls: 0.018 | Dom Loss: 0.095 | Dom Acc: 98.0%: 100%|██████████| 211/211 [00:08<00:00, 25.56it/s]


Epoch [3/20] Stats:
   Avg CL Loss: 0.0197
   Avg Domain Loss: 0.0830
   Source Acc: 99.03%
   Target Acc: 64.82%



Ep 4 | Cls: 0.036 | Dom Loss: 0.105 | Dom Acc: 97.5%: 100%|██████████| 211/211 [00:08<00:00, 25.82it/s]


Epoch [4/20] Stats:
   Avg CL Loss: 0.0195
   Avg Domain Loss: 0.0763
   Source Acc: 99.10%
   Target Acc: 65.00%



Ep 5 | Cls: 0.046 | Dom Loss: 0.147 | Dom Acc: 96.3%: 100%|██████████| 211/211 [00:08<00:00, 25.83it/s]


Epoch [5/20] Stats:
   Avg CL Loss: 0.0261
   Avg Domain Loss: 0.1111
   Source Acc: 99.11%
   Target Acc: 73.12%



Ep 6 | Cls: 0.023 | Dom Loss: 0.322 | Dom Acc: 87.5%: 100%|██████████| 211/211 [00:08<00:00, 26.16it/s]


Epoch [6/20] Stats:
   Avg CL Loss: 0.0408
   Avg Domain Loss: 0.2351
   Source Acc: 99.15%
   Target Acc: 68.67%



Ep 7 | Cls: 0.128 | Dom Loss: 0.207 | Dom Acc: 92.2%: 100%|██████████| 211/211 [00:08<00:00, 25.39it/s]


Epoch [7/20] Stats:
   Avg CL Loss: 0.0682
   Avg Domain Loss: 0.4642
   Source Acc: 98.35%
   Target Acc: 40.81%



Ep 8 | Cls: 0.023 | Dom Loss: 0.131 | Dom Acc: 97.3%: 100%|██████████| 211/211 [00:08<00:00, 25.77it/s]


Epoch [8/20] Stats:
   Avg CL Loss: 0.0362
   Avg Domain Loss: 0.1520
   Source Acc: 99.19%
   Target Acc: 62.70%



Ep 9 | Cls: 0.033 | Dom Loss: 0.201 | Dom Acc: 92.6%: 100%|██████████| 211/211 [00:08<00:00, 25.59it/s]


Epoch [9/20] Stats:
   Avg CL Loss: 0.0379
   Avg Domain Loss: 0.2473
   Source Acc: 98.69%
   Target Acc: 64.79%



Ep 10 | Cls: 0.040 | Dom Loss: 0.348 | Dom Acc: 88.3%: 100%|██████████| 211/211 [00:08<00:00, 25.56it/s]


Epoch [10/20] Stats:
   Avg CL Loss: 0.0471
   Avg Domain Loss: 0.3043
   Source Acc: 99.08%
   Target Acc: 78.96%



Ep 11 | Cls: 0.022 | Dom Loss: 0.320 | Dom Acc: 88.9%: 100%|██████████| 211/211 [00:08<00:00, 26.14it/s]


Epoch [11/20] Stats:
   Avg CL Loss: 0.0387
   Avg Domain Loss: 0.3153
   Source Acc: 98.72%
   Target Acc: 80.77%



Ep 12 | Cls: 0.061 | Dom Loss: 0.441 | Dom Acc: 81.4%: 100%|██████████| 211/211 [00:08<00:00, 26.04it/s]


Epoch [12/20] Stats:
   Avg CL Loss: 0.0476
   Avg Domain Loss: 0.3847
   Source Acc: 98.63%
   Target Acc: 78.38%



Ep 13 | Cls: 0.081 | Dom Loss: 0.438 | Dom Acc: 80.5%: 100%|██████████| 211/211 [00:08<00:00, 26.03it/s]


Epoch [13/20] Stats:
   Avg CL Loss: 0.0467
   Avg Domain Loss: 0.4165
   Source Acc: 98.86%
   Target Acc: 82.44%



Ep 14 | Cls: 0.092 | Dom Loss: 0.483 | Dom Acc: 75.6%: 100%|██████████| 211/211 [00:08<00:00, 26.12it/s]


Epoch [14/20] Stats:
   Avg CL Loss: 0.0443
   Avg Domain Loss: 0.4601
   Source Acc: 98.48%
   Target Acc: 73.10%



Ep 15 | Cls: 0.052 | Dom Loss: 0.548 | Dom Acc: 72.1%: 100%|██████████| 211/211 [00:08<00:00, 25.89it/s]


Epoch [15/20] Stats:
   Avg CL Loss: 0.0474
   Avg Domain Loss: 0.4831
   Source Acc: 98.72%
   Target Acc: 82.82%



Ep 16 | Cls: 0.023 | Dom Loss: 0.445 | Dom Acc: 80.3%: 100%|██████████| 211/211 [00:08<00:00, 25.03it/s]


Epoch [16/20] Stats:
   Avg CL Loss: 0.0451
   Avg Domain Loss: 0.5152
   Source Acc: 98.96%
   Target Acc: 85.67%



Ep 17 | Cls: 0.082 | Dom Loss: 0.677 | Dom Acc: 63.5%: 100%|██████████| 211/211 [00:08<00:00, 25.29it/s]


Epoch [17/20] Stats:
   Avg CL Loss: 0.0446
   Avg Domain Loss: 0.5066
   Source Acc: 98.49%
   Target Acc: 86.19%



Ep 18 | Cls: 0.040 | Dom Loss: 0.599 | Dom Acc: 67.4%: 100%|██████████| 211/211 [00:08<00:00, 26.04it/s]


Epoch [18/20] Stats:
   Avg CL Loss: 0.0431
   Avg Domain Loss: 0.5485
   Source Acc: 98.52%
   Target Acc: 85.58%



Ep 19 | Cls: 0.026 | Dom Loss: 0.652 | Dom Acc: 63.7%: 100%|██████████| 211/211 [00:08<00:00, 26.19it/s]


Epoch [19/20] Stats:
   Avg CL Loss: 0.0425
   Avg Domain Loss: 0.5653
   Source Acc: 98.82%
   Target Acc: 86.58%



Ep 20 | Cls: 0.066 | Dom Loss: 0.570 | Dom Acc: 70.7%: 100%|██████████| 211/211 [00:08<00:00, 25.64it/s]


Epoch [20/20] Stats:
   Avg CL Loss: 0.0393
   Avg Domain Loss: 0.5737
   Source Acc: 99.12%
   Target Acc: 85.33%

Run 1 Complete. Best Acc: 86.58%

  RUN 2/3 | SEED: 100
Seed set to 100
Loading pre-trained source weights from /content/drive/MyDrive/DA_Project/models/best_cnn_mnist_256_20_0.001.pth...
Loaded pre-trained source weights into DANN_CNN (features and class_classifier). Domain classifier initialized randomly.



Ep 1 | Cls: 0.045 | Dom Loss: 0.180 | Dom Acc: 99.8%: 100%|██████████| 211/211 [00:08<00:00, 25.92it/s]


Epoch [1/20] Stats:
   Avg CL Loss: 0.0197
   Avg Domain Loss: 0.3775
   Source Acc: 99.22%
   Target Acc: 43.41%



Ep 2 | Cls: 0.024 | Dom Loss: 0.093 | Dom Acc: 99.8%: 100%|██████████| 211/211 [00:08<00:00, 25.22it/s]


Epoch [2/20] Stats:
   Avg CL Loss: 0.0186
   Avg Domain Loss: 0.1234
   Source Acc: 99.17%
   Target Acc: 49.26%



Ep 3 | Cls: 0.020 | Dom Loss: 0.075 | Dom Acc: 99.2%: 100%|██████████| 211/211 [00:08<00:00, 25.63it/s]


Epoch [3/20] Stats:
   Avg CL Loss: 0.0174
   Avg Domain Loss: 0.0686
   Source Acc: 99.37%
   Target Acc: 55.78%



Ep 4 | Cls: 0.014 | Dom Loss: 0.053 | Dom Acc: 99.6%: 100%|██████████| 211/211 [00:08<00:00, 26.30it/s]


Epoch [4/20] Stats:
   Avg CL Loss: 0.0175
   Avg Domain Loss: 0.0587
   Source Acc: 99.35%
   Target Acc: 60.37%



Ep 5 | Cls: 0.012 | Dom Loss: 0.060 | Dom Acc: 99.6%: 100%|██████████| 211/211 [00:07<00:00, 26.49it/s]


Epoch [5/20] Stats:
   Avg CL Loss: 0.0203
   Avg Domain Loss: 0.0590
   Source Acc: 99.19%
   Target Acc: 64.93%



Ep 6 | Cls: 0.057 | Dom Loss: 0.236 | Dom Acc: 90.2%: 100%|██████████| 211/211 [00:08<00:00, 26.09it/s]


Epoch [6/20] Stats:
   Avg CL Loss: 0.0265
   Avg Domain Loss: 0.1222
   Source Acc: 98.92%
   Target Acc: 58.20%



Ep 7 | Cls: 0.020 | Dom Loss: 0.054 | Dom Acc: 99.4%: 100%|██████████| 211/211 [00:08<00:00, 25.62it/s]


Epoch [7/20] Stats:
   Avg CL Loss: 0.0487
   Avg Domain Loss: 0.2046
   Source Acc: 99.24%
   Target Acc: 34.28%



Ep 8 | Cls: 0.047 | Dom Loss: 0.051 | Dom Acc: 99.6%: 100%|██████████| 211/211 [00:08<00:00, 26.13it/s]


Epoch [8/20] Stats:
   Avg CL Loss: 0.0211
   Avg Domain Loss: 0.0715
   Source Acc: 99.29%
   Target Acc: 61.74%



Ep 9 | Cls: 0.051 | Dom Loss: 0.121 | Dom Acc: 97.1%: 100%|██████████| 211/211 [00:08<00:00, 25.40it/s]


Epoch [9/20] Stats:
   Avg CL Loss: 0.0261
   Avg Domain Loss: 0.1212
   Source Acc: 99.10%
   Target Acc: 69.20%



Ep 10 | Cls: 0.060 | Dom Loss: 0.170 | Dom Acc: 95.3%: 100%|██████████| 211/211 [00:08<00:00, 25.92it/s]


Epoch [10/20] Stats:
   Avg CL Loss: 0.0373
   Avg Domain Loss: 0.1892
   Source Acc: 99.12%
   Target Acc: 72.21%



Ep 11 | Cls: 0.019 | Dom Loss: 0.392 | Dom Acc: 82.8%: 100%|██████████| 211/211 [00:07<00:00, 26.44it/s]


Epoch [11/20] Stats:
   Avg CL Loss: 0.0477
   Avg Domain Loss: 0.3170
   Source Acc: 98.82%
   Target Acc: 79.01%



Ep 12 | Cls: 0.027 | Dom Loss: 0.467 | Dom Acc: 80.3%: 100%|██████████| 211/211 [00:08<00:00, 26.21it/s]


Epoch [12/20] Stats:
   Avg CL Loss: 0.0525
   Avg Domain Loss: 0.3932
   Source Acc: 98.73%
   Target Acc: 79.93%



Ep 13 | Cls: 0.114 | Dom Loss: 0.538 | Dom Acc: 75.4%: 100%|██████████| 211/211 [00:08<00:00, 25.56it/s]


Epoch [13/20] Stats:
   Avg CL Loss: 0.0527
   Avg Domain Loss: 0.4709
   Source Acc: 98.55%
   Target Acc: 81.14%



Ep 14 | Cls: 0.059 | Dom Loss: 0.527 | Dom Acc: 73.8%: 100%|██████████| 211/211 [00:08<00:00, 25.20it/s]


Epoch [14/20] Stats:
   Avg CL Loss: 0.0529
   Avg Domain Loss: 0.4981
   Source Acc: 98.63%
   Target Acc: 82.69%



Ep 15 | Cls: 0.015 | Dom Loss: 0.418 | Dom Acc: 83.2%: 100%|██████████| 211/211 [00:08<00:00, 25.27it/s]


Epoch [15/20] Stats:
   Avg CL Loss: 0.0516
   Avg Domain Loss: 0.5090
   Source Acc: 99.17%
   Target Acc: 84.94%



Ep 16 | Cls: 0.036 | Dom Loss: 0.592 | Dom Acc: 68.9%: 100%|██████████| 211/211 [00:08<00:00, 25.46it/s]


Epoch [16/20] Stats:
   Avg CL Loss: 0.0467
   Avg Domain Loss: 0.5114
   Source Acc: 98.52%
   Target Acc: 84.12%



Ep 17 | Cls: 0.067 | Dom Loss: 0.617 | Dom Acc: 68.0%: 100%|██████████| 211/211 [00:08<00:00, 25.65it/s]


Epoch [17/20] Stats:
   Avg CL Loss: 0.0433
   Avg Domain Loss: 0.5367
   Source Acc: 98.92%
   Target Acc: 87.92%



Ep 18 | Cls: 0.053 | Dom Loss: 0.423 | Dom Acc: 83.4%: 100%|██████████| 211/211 [00:07<00:00, 26.44it/s]


Epoch [18/20] Stats:
   Avg CL Loss: 0.0474
   Avg Domain Loss: 0.5851
   Source Acc: 98.78%
   Target Acc: 80.59%



Ep 19 | Cls: 0.027 | Dom Loss: 0.634 | Dom Acc: 67.8%: 100%|██████████| 211/211 [00:08<00:00, 25.95it/s]


Epoch [19/20] Stats:
   Avg CL Loss: 0.0464
   Avg Domain Loss: 0.6227
   Source Acc: 98.75%
   Target Acc: 84.76%



Ep 20 | Cls: 0.013 | Dom Loss: 0.535 | Dom Acc: 73.0%: 100%|██████████| 211/211 [00:08<00:00, 25.14it/s]


Epoch [20/20] Stats:
   Avg CL Loss: 0.0390
   Avg Domain Loss: 0.5946
   Source Acc: 99.12%
   Target Acc: 89.03%

Run 2 Complete. Best Acc: 89.03%

  RUN 3/3 | SEED: 2024
Seed set to 2024
Loading pre-trained source weights from /content/drive/MyDrive/DA_Project/models/best_cnn_mnist_256_20_0.001.pth...
Loaded pre-trained source weights into DANN_CNN (features and class_classifier). Domain classifier initialized randomly.



Ep 1 | Cls: 0.058 | Dom Loss: 0.201 | Dom Acc: 99.2%: 100%|██████████| 211/211 [00:08<00:00, 25.45it/s]


Epoch [1/20] Stats:
   Avg CL Loss: 0.0200
   Avg Domain Loss: 0.3617
   Source Acc: 98.78%
   Target Acc: 39.88%



Ep 2 | Cls: 0.002 | Dom Loss: 0.105 | Dom Acc: 99.2%: 100%|██████████| 211/211 [00:08<00:00, 25.74it/s]


Epoch [2/20] Stats:
   Avg CL Loss: 0.0214
   Avg Domain Loss: 0.1245
   Source Acc: 99.36%
   Target Acc: 56.97%



Ep 3 | Cls: 0.031 | Dom Loss: 0.085 | Dom Acc: 99.2%: 100%|██████████| 211/211 [00:08<00:00, 25.92it/s]


Epoch [3/20] Stats:
   Avg CL Loss: 0.0205
   Avg Domain Loss: 0.0890
   Source Acc: 99.15%
   Target Acc: 57.73%



Ep 4 | Cls: 0.022 | Dom Loss: 0.052 | Dom Acc: 99.4%: 100%|██████████| 211/211 [00:08<00:00, 25.71it/s]


Epoch [4/20] Stats:
   Avg CL Loss: 0.0217
   Avg Domain Loss: 0.0791
   Source Acc: 99.36%
   Target Acc: 64.34%



Ep 5 | Cls: 0.027 | Dom Loss: 0.186 | Dom Acc: 93.8%: 100%|██████████| 211/211 [00:08<00:00, 26.01it/s]


Epoch [5/20] Stats:
   Avg CL Loss: 0.0216
   Avg Domain Loss: 0.0938
   Source Acc: 98.93%
   Target Acc: 67.38%



Ep 6 | Cls: 0.058 | Dom Loss: 0.267 | Dom Acc: 90.8%: 100%|██████████| 211/211 [00:08<00:00, 25.44it/s]


Epoch [6/20] Stats:
   Avg CL Loss: 0.0427
   Avg Domain Loss: 0.2074
   Source Acc: 99.05%
   Target Acc: 73.77%



Ep 7 | Cls: 0.049 | Dom Loss: 0.272 | Dom Acc: 90.8%: 100%|██████████| 211/211 [00:08<00:00, 25.75it/s]


Epoch [7/20] Stats:
   Avg CL Loss: 0.0458
   Avg Domain Loss: 0.3099
   Source Acc: 98.96%
   Target Acc: 80.08%



Ep 8 | Cls: 0.035 | Dom Loss: 0.208 | Dom Acc: 93.6%: 100%|██████████| 211/211 [00:08<00:00, 25.89it/s]


Epoch [8/20] Stats:
   Avg CL Loss: 0.0523
   Avg Domain Loss: 0.3111
   Source Acc: 99.00%
   Target Acc: 62.59%



Ep 9 | Cls: 0.122 | Dom Loss: 0.315 | Dom Acc: 88.7%: 100%|██████████| 211/211 [00:08<00:00, 25.57it/s]


Epoch [9/20] Stats:
   Avg CL Loss: 0.0433
   Avg Domain Loss: 0.2629
   Source Acc: 98.91%
   Target Acc: 73.40%



Ep 10 | Cls: 0.049 | Dom Loss: 0.444 | Dom Acc: 81.4%: 100%|██████████| 211/211 [00:08<00:00, 25.83it/s]


Epoch [10/20] Stats:
   Avg CL Loss: 0.0502
   Avg Domain Loss: 0.3714
   Source Acc: 98.82%
   Target Acc: 82.54%



Ep 11 | Cls: 0.081 | Dom Loss: 0.411 | Dom Acc: 81.2%: 100%|██████████| 211/211 [00:07<00:00, 26.51it/s]


Epoch [11/20] Stats:
   Avg CL Loss: 0.0476
   Avg Domain Loss: 0.4080
   Source Acc: 98.98%
   Target Acc: 82.36%



Ep 12 | Cls: 0.025 | Dom Loss: 0.363 | Dom Acc: 88.1%: 100%|██████████| 211/211 [00:08<00:00, 25.31it/s]


Epoch [12/20] Stats:
   Avg CL Loss: 0.0430
   Avg Domain Loss: 0.4444
   Source Acc: 99.01%
   Target Acc: 79.76%



Ep 13 | Cls: 0.025 | Dom Loss: 0.456 | Dom Acc: 77.9%: 100%|██████████| 211/211 [00:08<00:00, 25.94it/s]


Epoch [13/20] Stats:
   Avg CL Loss: 0.0437
   Avg Domain Loss: 0.4821
   Source Acc: 98.98%
   Target Acc: 83.21%



Ep 14 | Cls: 0.017 | Dom Loss: 0.419 | Dom Acc: 81.2%: 100%|██████████| 211/211 [00:08<00:00, 25.53it/s]


Epoch [14/20] Stats:
   Avg CL Loss: 0.0428
   Avg Domain Loss: 0.4701
   Source Acc: 99.22%
   Target Acc: 85.73%



Ep 15 | Cls: 0.025 | Dom Loss: 0.467 | Dom Acc: 78.3%: 100%|██████████| 211/211 [00:08<00:00, 25.94it/s]


Epoch [15/20] Stats:
   Avg CL Loss: 0.0446
   Avg Domain Loss: 0.5092
   Source Acc: 98.61%
   Target Acc: 87.33%



Ep 16 | Cls: 0.035 | Dom Loss: 0.673 | Dom Acc: 63.5%: 100%|██████████| 211/211 [00:08<00:00, 26.12it/s]


Epoch [16/20] Stats:
   Avg CL Loss: 0.0420
   Avg Domain Loss: 0.5216
   Source Acc: 98.73%
   Target Acc: 89.41%



Ep 17 | Cls: 0.110 | Dom Loss: 0.577 | Dom Acc: 70.1%: 100%|██████████| 211/211 [00:08<00:00, 26.33it/s]


Epoch [17/20] Stats:
   Avg CL Loss: 0.0419
   Avg Domain Loss: 0.5677
   Source Acc: 99.09%
   Target Acc: 91.51%



Ep 18 | Cls: 0.042 | Dom Loss: 0.628 | Dom Acc: 67.6%: 100%|██████████| 211/211 [00:08<00:00, 25.73it/s]


Epoch [18/20] Stats:
   Avg CL Loss: 0.0423
   Avg Domain Loss: 0.5823
   Source Acc: 98.87%
   Target Acc: 88.43%



Ep 19 | Cls: 0.032 | Dom Loss: 0.544 | Dom Acc: 70.9%: 100%|██████████| 211/211 [00:08<00:00, 25.74it/s]


Epoch [19/20] Stats:
   Avg CL Loss: 0.0429
   Avg Domain Loss: 0.6009
   Source Acc: 99.26%
   Target Acc: 90.46%



Ep 20 | Cls: 0.021 | Dom Loss: 0.614 | Dom Acc: 69.3%: 100%|██████████| 211/211 [00:08<00:00, 25.32it/s]


Epoch [20/20] Stats:
   Avg CL Loss: 0.0389
   Avg Domain Loss: 0.5817
   Source Acc: 99.13%
   Target Acc: 89.52%

Run 3 Complete. Best Acc: 91.51%

 ROBUSTNESS EXPERIMENT COMPLETE
   Individual Runs: [86.57777777777778, 89.03333333333333, 91.5111111111111]
   Final Result: 89.04% ± 2.01%


The results were outstanding and validated the stability of this approach. Across the three runs, we achieved best target accuracies of **86.58%**, **89.03%**, and **91.51%**, resulting in a **mean of 89.04% ± 2.01%**. This performance is a decisive improvement over our internal distance-based (MMD) baseline (79.36%) and significantly outperforms the original DANN paper’s benchmark of 76.1% ± 1.8%. Source accuracy remained robust throughout (>98.5%), confirming that the adaptation did not degrade the primary task performance.

Training dynamics consistently showed characteristic adversarial fluctuations followed by strong recovery. While target accuracy often dipped in early epochs (epochs 4–7), it reliably climbed into the high 80s by the later stages. The domain classifier accuracy settled between **69% and 77%** (with BCE loss $\approx$ 0.57–0.60), rather than the theoretical optimum of 50%. This confirms our "Sufficient Alignment" hypothesis: the feature extractor successfully aligned the semantic digit features to drive state-of-the-art classification, even while retaining enough domain-specific noise to allow the discriminator to perform better than random guessing.

This robustness check solidifies that matching batch statistics (256->256) combined with a weak fixed adversary is a highly effective and reproducible recipe for Domain Adaptation on MNIST-M.

### Experiment 2: Stability Analysis — α and Learning Rate Annealing

While Experiment 1 yielded high peak accuracy, the training dynamics exhibited significant instability, with target accuracy dropping as low as ~40% during epochs 4–7 before recovering. To mitigate this and test if a more gradual optimization path yields better convergence, we adopt the specific annealing schedules proposed in the original DANN paper.

The objective is to introduce a "warm-up" phase where the model prioritizes learning basic semantic features before the adversarial penalty becomes significant. We implement two specific schedules:

**1. Annealing $\alpha$ (Adversarial Strength):**
Instead of a fixed $\alpha = 0.25$, we schedule the gradient reversal strength to ramp smoothly from 0 to 1.0 over the course of training:
$$
\alpha_p = \frac{2}{1 + \exp(-10 \cdot p)} - 1
$$
where $p \in [0, 1]$ represents the training progress (current epoch / total epochs). This ensures early updates are effectively non-adversarial ($\alpha \approx 0$), allowing the feature extractor to stabilize, before ramping to full strength ($\alpha \to 1.0$) to force maximal domain confusion.

**2. Annealing Learning Rate:**
We implement the inverse decay schedule to gradually reduce the step size as training progresses:
$$
\text{LR}_p = \frac{\text{LR}_0}{(1 + 10 \cdot p)^{0.75}}
$$
where $\text{LR}_0 = 10^{-3}$. This facilitates fast initial convergence while preventing the model from oscillating or "over-correcting" when the adversarial constraint becomes strong in later epochs.

**Experimental Configuration:**
* **Initialization:** Pre-trained Source Model (Batch 256)
* **Batch Size:** 256
* **Optimizer:** AdamW
* **Learning Rate:** Annealed ($10^{-3} \to \sim 2\times 10^{-4}$)
* **Adversarial $\alpha$:** Annealed ($0.0 \to 1.0$)

In [11]:
LEARNING_RATE = 1e-3
BATCH_SIZE = 256
max_alpha = 1.0

# Target location of our saved model
models_path = "/content/drive/MyDrive/DA_Project/models/"
best_model_file_path = os.path.join(models_path, f"best_cnn_mnist_{BATCH_SIZE}_{NUM_EPOCHS}_{LEARNING_RATE}.pth")

# Initialize DANN
dann_model = DANN_CNN(num_classes=10).to(device)
if os.path.exists(best_model_file_path):
    print(f"Loading pre-trained source weights from {best_model_file_path}...")
    pretrained_state_dict = torch.load(best_model_file_path, map_location=device)

    # Create a new state_dict for the DANN model, which we'll populate
    dann_state_dict = dann_model.state_dict()

    # Copy weights for the feature extractor
    for k, v in pretrained_state_dict.items():
        if k.startswith('features.'):
            dann_state_dict[k] = v
        elif k.startswith('classifier.'):
            # Rename 'classifier' keys to 'class_classifier' to match DANN_CNN
            new_key = k.replace('classifier.', 'class_classifier.')
            dann_state_dict[new_key] = v

    # Load the modified state_dict. strict=False is used because 'domain_classifier'
    # keys will be missing from the loaded pretrained_state_dict, which is intended.
    dann_model.load_state_dict(dann_state_dict, strict=False)
    print("Loaded pre-trained source weights into DANN_CNN (features and class_classifier). Domain classifier initialized randomly.\n")
else:
    print("Pre-trained weights not found. Starting DANN_CNN from scratch.")

criterion = nn.CrossEntropyLoss()
criterion_domain = nn.BCEWithLogitsLoss()
optimizer = optim.AdamW(dann_model.parameters(), lr=LEARNING_RATE)

best_target_acc = 0.0

print(f" Starting DANN Training (Adversarial)...")
start_time = time.time()

for epoch in range(NUM_EPOCHS):
    dann_model.train()

    # Stats tracking
    total_loss, total_domain_loss, total_class_loss = 0.0, 0.0, 0.0
    total_samples = 0

    # Schedule alpha
    progress = epoch / NUM_EPOCHS
    annealing_factor = 2 / (1 + np.exp(-10 * progress)) - 1
    alpha = max_alpha * annealing_factor

    # Schedule learning rate
    NEW_LEARNING_RATE = LEARNING_RATE / ((1 + 10 * progress) ** 0.75)
    for param_group in optimizer.param_groups:
        param_group['lr'] = NEW_LEARNING_RATE

    # Zip loaders
    min_len = min(len(source_train_loader), len(target_train_loader))
    loader_zip = tqdm(zip(source_train_loader, target_train_loader), total=min_len)

    for (source_imgs, source_labels), (target_imgs, _) in loader_zip:
        # Setup
        source_imgs, source_labels = source_imgs.to(device), source_labels.to(device)
        target_imgs = target_imgs.to(device)
        batch_size = source_imgs.size(0)    # we do this once again to make sure of the batch size

        # Handle uneven batches
        if source_imgs.size(0) != target_imgs.size(0): continue

        # Create Domain Labels: Source = 0, Target = 1
        domain_label_s = torch.zeros(batch_size,1).float().to(device)
        domain_label_t = torch.ones(batch_size,1).float().to(device)

        optimizer.zero_grad()

        # Forward Pass (Source Data)
        # We calculate BOTH Class Loss and Domain Loss
        class_logits_s, domain_logits_s = dann_model(source_imgs, alpha)
        loss_class = criterion(class_logits_s, source_labels)
        loss_domain_s = criterion_domain(domain_logits_s, domain_label_s)

        # Forward Pass (Target Data)
        # We ONLY calculate Domain Loss (we don't have class labels)
        _, domain_logits_t = dann_model(target_imgs, alpha)
        loss_domain_t = criterion_domain(domain_logits_t, domain_label_t)

        # Backward Pass (Optimization)
        # The GRL inside the model handles the sign flipping automatically!
        loss_domain = (loss_domain_s + loss_domain_t)/2
        loss = loss_class + loss_domain

        loss.backward()
        optimizer.step()

        # Stats
        total_loss += loss.item()
        total_class_loss += loss_class.item()
        total_domain_loss += loss_domain.item()
        total_samples += 1

        with torch.no_grad():
            prob_s = torch.sigmoid(domain_logits_s).mean().item()
            prob_t = torch.sigmoid(domain_logits_t).mean().item()

            # Accuracy checks

            # Source is correctly classified if logit < 0 (Prob < 0.5)
            acc_s = (domain_logits_s < 0).float().mean().item() * 100
            # Target is correctly classified if logit > 0 (Prob > 0.5)
            acc_t = (domain_logits_t > 0).float().mean().item() * 100

            domain_acc = (acc_s + acc_t) / 2

        # Logging
        loader_zip.set_description(
            f"Ep {epoch+1} | Cls: {loss_class.item():.3f} | "
            f"Dom Loss: {loss_domain.item():.3f} | "
            f"Dom Acc: {domain_acc:.1f}%"
        )

        # Average classification loss and domain loss
        avg_class_loss = total_class_loss / total_samples
        avg_domain_loss = total_domain_loss / total_samples

    # Validation
    # Evaluate on Target (Did adaptation work?)
    target_acc = evaluate_accuracy(dann_model, target_test_loader, device)

    # Evaluate on Source (Did we preserve original knowledge?)
    source_acc = evaluate_accuracy(dann_model, source_test_loader, device)

    print(f"Epoch [{epoch+1}/{NUM_EPOCHS}] Stats:")
    print(f"   Avg CL Loss: {avg_class_loss:.4f}")
    print(f"   Avg Domain Loss: {avg_domain_loss:.4f}")
    print(f"   Source Acc: {source_acc:.2f}%")
    print(f"   Target Acc: {target_acc:.2f}%\n")

    # Checkpointing
    if target_acc > best_target_acc:
        best_target_acc = target_acc
        best_model_wts = copy.deepcopy(dann_model.state_dict())

total_time = time.time() - start_time

# Save the best model weights
target_file_path_best = os.path.join(models_path, f"best_DANN_Exp2{BATCH_SIZE}_{NUM_EPOCHS}_{LEARNING_RATE}annealed_{max_alpha}annealed.pth")
torch.save(best_model_wts, target_file_path_best)
print(f"Best model weights saved to {target_file_path_best}")
print(f"\nAdaptation Complete in {total_time:.1f}s Best Target Accuracy: {best_target_acc:.2f}%")

Loading pre-trained source weights from /content/drive/MyDrive/DA_Project/models/best_cnn_mnist_256_20_0.001.pth...
Loaded pre-trained source weights into DANN_CNN (features and class_classifier). Domain classifier initialized randomly.

 Starting DANN Training (Adversarial)...


Ep 1 | Cls: 0.049 | Dom Loss: 0.085 | Dom Acc: 100.0%: 100%|██████████| 211/211 [00:08<00:00, 25.71it/s]


Epoch [1/20] Stats:
   Avg CL Loss: 0.0159
   Avg Domain Loss: 0.3140
   Source Acc: 98.76%
   Target Acc: 50.20%



Ep 2 | Cls: 0.024 | Dom Loss: 0.113 | Dom Acc: 99.6%: 100%|██████████| 211/211 [00:08<00:00, 25.70it/s]


Epoch [2/20] Stats:
   Avg CL Loss: 0.0167
   Avg Domain Loss: 0.1193
   Source Acc: 99.39%
   Target Acc: 44.69%



Ep 3 | Cls: 0.006 | Dom Loss: 0.131 | Dom Acc: 98.4%: 100%|██████████| 211/211 [00:08<00:00, 25.75it/s]


Epoch [3/20] Stats:
   Avg CL Loss: 0.0279
   Avg Domain Loss: 0.1485
   Source Acc: 99.18%
   Target Acc: 49.74%



Ep 4 | Cls: 0.081 | Dom Loss: 0.345 | Dom Acc: 86.1%: 100%|██████████| 211/211 [00:08<00:00, 25.23it/s]


Epoch [4/20] Stats:
   Avg CL Loss: 0.0619
   Avg Domain Loss: 0.2634
   Source Acc: 98.14%
   Target Acc: 56.29%



Ep 5 | Cls: 0.073 | Dom Loss: 0.373 | Dom Acc: 86.7%: 100%|██████████| 211/211 [00:08<00:00, 25.93it/s]


Epoch [5/20] Stats:
   Avg CL Loss: 0.0914
   Avg Domain Loss: 0.4556
   Source Acc: 98.62%
   Target Acc: 69.18%



Ep 6 | Cls: 0.122 | Dom Loss: 0.515 | Dom Acc: 75.8%: 100%|██████████| 211/211 [00:08<00:00, 26.36it/s]


Epoch [6/20] Stats:
   Avg CL Loss: 0.0877
   Avg Domain Loss: 0.4649
   Source Acc: 98.07%
   Target Acc: 74.66%



Ep 7 | Cls: 0.137 | Dom Loss: 0.638 | Dom Acc: 67.2%: 100%|██████████| 211/211 [00:08<00:00, 26.26it/s]


Epoch [7/20] Stats:
   Avg CL Loss: 0.0925
   Avg Domain Loss: 0.5228
   Source Acc: 96.63%
   Target Acc: 66.84%



Ep 8 | Cls: 0.086 | Dom Loss: 0.536 | Dom Acc: 74.2%: 100%|██████████| 211/211 [00:08<00:00, 25.82it/s]


Epoch [8/20] Stats:
   Avg CL Loss: 0.0920
   Avg Domain Loss: 0.5370
   Source Acc: 98.15%
   Target Acc: 75.31%



Ep 9 | Cls: 0.106 | Dom Loss: 0.486 | Dom Acc: 79.5%: 100%|██████████| 211/211 [00:08<00:00, 26.13it/s]


Epoch [9/20] Stats:
   Avg CL Loss: 0.0857
   Avg Domain Loss: 0.5234
   Source Acc: 98.67%
   Target Acc: 78.93%



Ep 10 | Cls: 0.108 | Dom Loss: 0.508 | Dom Acc: 77.7%: 100%|██████████| 211/211 [00:08<00:00, 25.83it/s]


Epoch [10/20] Stats:
   Avg CL Loss: 0.0811
   Avg Domain Loss: 0.4973
   Source Acc: 98.59%
   Target Acc: 82.02%



Ep 11 | Cls: 0.164 | Dom Loss: 0.512 | Dom Acc: 77.1%: 100%|██████████| 211/211 [00:08<00:00, 26.09it/s]


Epoch [11/20] Stats:
   Avg CL Loss: 0.0771
   Avg Domain Loss: 0.4960
   Source Acc: 96.60%
   Target Acc: 65.17%



Ep 12 | Cls: 0.066 | Dom Loss: 0.534 | Dom Acc: 74.2%: 100%|██████████| 211/211 [00:07<00:00, 26.40it/s]


Epoch [12/20] Stats:
   Avg CL Loss: 0.0899
   Avg Domain Loss: 0.5444
   Source Acc: 98.20%
   Target Acc: 79.26%



Ep 13 | Cls: 0.036 | Dom Loss: 0.526 | Dom Acc: 75.0%: 100%|██████████| 211/211 [00:08<00:00, 25.81it/s]


Epoch [13/20] Stats:
   Avg CL Loss: 0.0765
   Avg Domain Loss: 0.5338
   Source Acc: 98.23%
   Target Acc: 79.11%



Ep 14 | Cls: 0.159 | Dom Loss: 0.601 | Dom Acc: 67.8%: 100%|██████████| 211/211 [00:08<00:00, 25.49it/s]


Epoch [14/20] Stats:
   Avg CL Loss: 0.0823
   Avg Domain Loss: 0.5437
   Source Acc: 97.66%
   Target Acc: 78.70%



Ep 15 | Cls: 0.062 | Dom Loss: 0.537 | Dom Acc: 74.2%: 100%|██████████| 211/211 [00:08<00:00, 25.36it/s]


Epoch [15/20] Stats:
   Avg CL Loss: 0.0713
   Avg Domain Loss: 0.5137
   Source Acc: 98.84%
   Target Acc: 75.90%



Ep 16 | Cls: 0.064 | Dom Loss: 0.550 | Dom Acc: 70.7%: 100%|██████████| 211/211 [00:08<00:00, 25.40it/s]


Epoch [16/20] Stats:
   Avg CL Loss: 0.0698
   Avg Domain Loss: 0.5350
   Source Acc: 97.85%
   Target Acc: 76.59%



Ep 17 | Cls: 0.085 | Dom Loss: 0.656 | Dom Acc: 64.5%: 100%|██████████| 211/211 [00:08<00:00, 25.71it/s]


Epoch [17/20] Stats:
   Avg CL Loss: 0.0829
   Avg Domain Loss: 0.5628
   Source Acc: 98.31%
   Target Acc: 76.91%



Ep 18 | Cls: 0.152 | Dom Loss: 0.601 | Dom Acc: 70.3%: 100%|██████████| 211/211 [00:08<00:00, 26.01it/s]


Epoch [18/20] Stats:
   Avg CL Loss: 0.0825
   Avg Domain Loss: 0.5694
   Source Acc: 97.80%
   Target Acc: 74.96%



Ep 19 | Cls: 0.080 | Dom Loss: 0.569 | Dom Acc: 68.8%: 100%|██████████| 211/211 [00:08<00:00, 26.36it/s]


Epoch [19/20] Stats:
   Avg CL Loss: 0.0746
   Avg Domain Loss: 0.5292
   Source Acc: 98.15%
   Target Acc: 81.26%



Ep 20 | Cls: 0.166 | Dom Loss: 0.562 | Dom Acc: 71.9%: 100%|██████████| 211/211 [00:08<00:00, 26.06it/s]


Epoch [20/20] Stats:
   Avg CL Loss: 0.0857
   Avg Domain Loss: 0.5701
   Source Acc: 97.86%
   Target Acc: 75.03%

Best model weights saved to /content/drive/MyDrive/DA_Project/models/best_DANN_Exp2256_20_0.001annealed_1.0annealed.pth

Adaptation Complete in 212.9s Best Target Accuracy: 82.02%


### Takeaway Experiment 2: Annealing for Stability

In this experiment, we adopted the standard DANN optimization strategy: initializing with Batch 256 weights and annealing the adversarial strength $\alpha$ from 0 to 1, coupled with a decaying learning rate. This configuration achieved a best target accuracy of 82.02%.
While this performance confirms the effectiveness of adversarial adaptation—surpassing the 79.36% distance-based benchmark—it falls significantly short of the ~89% achieved by our 'Fixed Weak Alpha' strategy. These results suggest that while the annealing 'warm-up' is beneficial in early epochs, allowing the adversarial strength to ramp up to $\alpha \approx 1.0$ forces the discriminator to become too aggressive. This over-correction appears to degrade the semantic features, confirming that for MNIST-M, a constant, gentle adversarial pressure ($\alpha=0.25$) yields better generalization than the theoretically standard schedule of forcing perfect domain confusion.

In [10]:
SEEDS = [42, 100, 2024]
LEARNING_RATE = 1e-3
BATCH_SIZE = 256
max_alpha = 1.0

# Target location of our saved model
models_path = "/content/drive/MyDrive/DA_Project/models/"
best_model_file_path = os.path.join(models_path, f"best_cnn_mnist_{BATCH_SIZE}_{NUM_EPOCHS}_{LEARNING_RATE}.pth")

# Store final results
final_accuracies = []

print(f"   Starting Robustness Experiment on {len(SEEDS)} seeds...")
print(f"   Config: BS={BATCH_SIZE}, LR={LEARNING_RATE}, Alpha={max_alpha}")

start_time = time.time()

for run_idx, seed in enumerate(SEEDS):
    print(f"\n" + "="*40)
    print(f"  RUN {run_idx+1}/{len(SEEDS)} | SEED: {seed}")
    print("="*40)

    # set seed
    set_seed(seed)

    # Re-Initialize DataLoaders
    source_train_loader = DataLoader(train_subset, batch_size=BATCH_SIZE, shuffle=True, num_workers=NUM_WORKERS, pin_memory=True)
    source_test_loader = DataLoader(mnist_test, batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS, pin_memory=True)
    target_train_loader = DataLoader(mnistm_train, batch_size=BATCH_SIZE, shuffle=True, num_workers=NUM_WORKERS, pin_memory=True)
    target_test_loader = DataLoader(mnistm_test, batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS, pin_memory=True)


    # Initialize DANN
    dann_model = DANN_CNN(num_classes=10).to(device)
    if os.path.exists(best_model_file_path):
        print(f"Loading pre-trained source weights from {best_model_file_path}...")
        pretrained_state_dict = torch.load(best_model_file_path, map_location=device)

        # Create a new state_dict for the DANN model, which we'll populate
        dann_state_dict = dann_model.state_dict()

        # Copy weights for the feature extractor
        for k, v in pretrained_state_dict.items():
            if k.startswith('features.'):
                dann_state_dict[k] = v
            elif k.startswith('classifier.'):
                # Rename 'classifier' keys to 'class_classifier' to match DANN_CNN
                new_key = k.replace('classifier.', 'class_classifier.')
                dann_state_dict[new_key] = v

        # Load the modified state_dict. strict=False is used because 'domain_classifier'
        # keys will be missing from the loaded pretrained_state_dict, which is intended.
        dann_model.load_state_dict(dann_state_dict, strict=False)
        print("Loaded pre-trained source weights into DANN_CNN (features and class_classifier). Domain classifier initialized randomly.\n")
    else:
        print("Pre-trained weights not found. Starting DANN_CNN from scratch.")

    criterion = nn.CrossEntropyLoss()
    criterion_domain = nn.BCEWithLogitsLoss()
    optimizer = optim.AdamW(dann_model.parameters(), lr=LEARNING_RATE)
    best_target_acc = 0.0

    for epoch in range(NUM_EPOCHS):
        dann_model.train()

        # Stats tracking
        total_loss, total_domain_loss, total_class_loss = 0.0, 0.0, 0.0
        total_samples = 0

        # Schedule alpha
        progress = epoch / NUM_EPOCHS
        annealing_factor = 2 / (1 + np.exp(-10 * progress)) - 1
        alpha = max_alpha * annealing_factor

        # Schedule learning rate
        NEW_LEARNING_RATE = LEARNING_RATE / ((1 + 10 * progress) ** 0.75)
        for param_group in optimizer.param_groups:
            param_group['lr'] = NEW_LEARNING_RATE

        # Zip loaders
        min_len = min(len(source_train_loader), len(target_train_loader))
        loader_zip = tqdm(zip(source_train_loader, target_train_loader), total=min_len)

        for (source_imgs, source_labels), (target_imgs, _) in loader_zip:
            # Setup
            source_imgs, source_labels = source_imgs.to(device), source_labels.to(device)
            target_imgs = target_imgs.to(device)
            batch_size = source_imgs.size(0)    # we do this once again to make sure of the batch size

            # Handle uneven batches
            if source_imgs.size(0) != target_imgs.size(0): continue

            # Create Domain Labels: Source = 0, Target = 1
            domain_label_s = torch.zeros(batch_size,1).float().to(device)
            domain_label_t = torch.ones(batch_size,1).float().to(device)

            optimizer.zero_grad()

            # Forward Pass (Source Data)
            # We calculate BOTH Class Loss and Domain Loss
            class_logits_s, domain_logits_s = dann_model(source_imgs, alpha)
            loss_class = criterion(class_logits_s, source_labels)
            loss_domain_s = criterion_domain(domain_logits_s, domain_label_s)

            # Forward Pass (Target Data)
            # We ONLY calculate Domain Loss (we don't have class labels)
            _, domain_logits_t = dann_model(target_imgs, alpha)
            loss_domain_t = criterion_domain(domain_logits_t, domain_label_t)

            # Backward Pass (Optimization)
            # The GRL inside the model handles the sign flipping automatically!
            loss_domain = (loss_domain_s + loss_domain_t)/2
            loss = loss_class + loss_domain

            loss.backward()
            optimizer.step()

            # Stats
            total_loss += loss.item()
            total_class_loss += loss_class.item()
            total_domain_loss += loss_domain.item()
            total_samples += 1

            with torch.no_grad():
                prob_s = torch.sigmoid(domain_logits_s).mean().item()
                prob_t = torch.sigmoid(domain_logits_t).mean().item()

                # Accuracy checks

                # Source is correctly classified if logit < 0 (Prob < 0.5)
                acc_s = (domain_logits_s < 0).float().mean().item() * 100
                # Target is correctly classified if logit > 0 (Prob > 0.5)
                acc_t = (domain_logits_t > 0).float().mean().item() * 100

                domain_acc = (acc_s + acc_t) / 2

            # Logging
            loader_zip.set_description(
                f"Ep {epoch+1} | Cls: {loss_class.item():.3f} | "
                f"Dom Loss: {loss_domain.item():.3f} | "
                f"Dom Acc: {domain_acc:.1f}%"
            )

            # Average classification loss and domain loss
            avg_class_loss = total_class_loss / total_samples
            avg_domain_loss = total_domain_loss / total_samples

        # Validation
        # Evaluate on Target (Did adaptation work?)
        target_acc = evaluate_accuracy(dann_model, target_test_loader, device)

        # Evaluate on Source (Did we preserve original knowledge?)
        source_acc = evaluate_accuracy(dann_model, source_test_loader, device)

        print(f"Epoch [{epoch+1}/{NUM_EPOCHS}] Stats:")
        print(f"   Avg CL Loss: {avg_class_loss:.4f}")
        print(f"   Avg Domain Loss: {avg_domain_loss:.4f}")
        print(f"   Source Acc: {source_acc:.2f}%")
        print(f"   Target Acc: {target_acc:.2f}%\n")

        # Checkpointing
        if target_acc > best_target_acc:
            best_target_acc = target_acc
            best_model_wts = copy.deepcopy(dann_model.state_dict())

    print(f"Run {run_idx+1} Complete. Best Acc: {best_target_acc:.2f}%")
    final_accuracies.append(best_target_acc)

total_time = time.time() - start_time
mean_acc = np.mean(final_accuracies)
std_acc = np.std(final_accuracies)
print(f"\n ROBUSTNESS EXPERIMENT COMPLETE WITH ANNEALING")
print(f"   Individual Runs: {final_accuracies}")
print(f"   Final Result: {mean_acc:.2f}% ± {std_acc:.2f}%")

   Starting Robustness Experiment on 3 seeds...
   Config: BS=256, LR=0.001, Alpha=0.24996257688624474

  RUN 1/3 | SEED: 42
Seed set to 42
Loading pre-trained source weights from /content/drive/MyDrive/DA_Project/models/best_cnn_mnist_256_20_0.001.pth...
Loaded pre-trained source weights into DANN_CNN (features and class_classifier). Domain classifier initialized randomly.



Ep 1 | Cls: 0.002 | Dom Loss: 0.076 | Dom Acc: 100.0%: 100%|██████████| 211/211 [00:08<00:00, 25.11it/s]


Epoch [1/20] Stats:
   Avg CL Loss: 0.0177
   Avg Domain Loss: 0.2931
   Source Acc: 99.31%
   Target Acc: 49.29%



Ep 2 | Cls: 0.041 | Dom Loss: 0.089 | Dom Acc: 99.6%: 100%|██████████| 211/211 [00:08<00:00, 25.43it/s]


Epoch [2/20] Stats:
   Avg CL Loss: 0.0150
   Avg Domain Loss: 0.1151
   Source Acc: 99.23%
   Target Acc: 44.71%



Ep 3 | Cls: 0.017 | Dom Loss: 0.256 | Dom Acc: 93.9%: 100%|██████████| 211/211 [00:08<00:00, 26.02it/s]


Epoch [3/20] Stats:
   Avg CL Loss: 0.0325
   Avg Domain Loss: 0.1703
   Source Acc: 98.70%
   Target Acc: 58.84%



Ep 4 | Cls: 0.039 | Dom Loss: 0.215 | Dom Acc: 94.9%: 100%|██████████| 211/211 [00:08<00:00, 25.68it/s]


Epoch [4/20] Stats:
   Avg CL Loss: 0.0794
   Avg Domain Loss: 0.3658
   Source Acc: 99.08%
   Target Acc: 48.74%



Ep 5 | Cls: 0.097 | Dom Loss: 0.366 | Dom Acc: 88.7%: 100%|██████████| 211/211 [00:08<00:00, 26.08it/s]


Epoch [5/20] Stats:
   Avg CL Loss: 0.0848
   Avg Domain Loss: 0.3714
   Source Acc: 98.55%
   Target Acc: 63.27%



Ep 6 | Cls: 0.079 | Dom Loss: 0.427 | Dom Acc: 83.2%: 100%|██████████| 211/211 [00:08<00:00, 25.58it/s]


Epoch [6/20] Stats:
   Avg CL Loss: 0.0879
   Avg Domain Loss: 0.4536
   Source Acc: 98.08%
   Target Acc: 71.76%



Ep 7 | Cls: 0.132 | Dom Loss: 0.444 | Dom Acc: 84.2%: 100%|██████████| 211/211 [00:08<00:00, 25.58it/s]


Epoch [7/20] Stats:
   Avg CL Loss: 0.0935
   Avg Domain Loss: 0.4899
   Source Acc: 98.87%
   Target Acc: 71.23%



Ep 8 | Cls: 0.130 | Dom Loss: 0.593 | Dom Acc: 69.5%: 100%|██████████| 211/211 [00:08<00:00, 25.62it/s]


Epoch [8/20] Stats:
   Avg CL Loss: 0.0906
   Avg Domain Loss: 0.4965
   Source Acc: 98.45%
   Target Acc: 67.64%



Ep 9 | Cls: 0.024 | Dom Loss: 0.481 | Dom Acc: 81.1%: 100%|██████████| 211/211 [00:08<00:00, 26.21it/s]


Epoch [9/20] Stats:
   Avg CL Loss: 0.0862
   Avg Domain Loss: 0.5000
   Source Acc: 98.23%
   Target Acc: 72.17%



Ep 10 | Cls: 0.033 | Dom Loss: 0.464 | Dom Acc: 80.7%: 100%|██████████| 211/211 [00:08<00:00, 25.64it/s]


Epoch [10/20] Stats:
   Avg CL Loss: 0.0859
   Avg Domain Loss: 0.5127
   Source Acc: 98.86%
   Target Acc: 74.17%



Ep 11 | Cls: 0.058 | Dom Loss: 0.544 | Dom Acc: 73.8%: 100%|██████████| 211/211 [00:07<00:00, 26.43it/s]


Epoch [11/20] Stats:
   Avg CL Loss: 0.0890
   Avg Domain Loss: 0.5484
   Source Acc: 98.32%
   Target Acc: 75.53%



Ep 12 | Cls: 0.101 | Dom Loss: 0.490 | Dom Acc: 76.8%: 100%|██████████| 211/211 [00:08<00:00, 25.78it/s]


Epoch [12/20] Stats:
   Avg CL Loss: 0.0869
   Avg Domain Loss: 0.5054
   Source Acc: 98.48%
   Target Acc: 68.52%



Ep 13 | Cls: 0.102 | Dom Loss: 0.567 | Dom Acc: 70.9%: 100%|██████████| 211/211 [00:08<00:00, 25.87it/s]


Epoch [13/20] Stats:
   Avg CL Loss: 0.0912
   Avg Domain Loss: 0.5713
   Source Acc: 98.52%
   Target Acc: 76.90%



Ep 14 | Cls: 0.073 | Dom Loss: 0.500 | Dom Acc: 79.3%: 100%|██████████| 211/211 [00:08<00:00, 25.27it/s]


Epoch [14/20] Stats:
   Avg CL Loss: 0.0830
   Avg Domain Loss: 0.5396
   Source Acc: 98.83%
   Target Acc: 76.87%



Ep 15 | Cls: 0.091 | Dom Loss: 0.590 | Dom Acc: 69.1%: 100%|██████████| 211/211 [00:08<00:00, 25.93it/s]


Epoch [15/20] Stats:
   Avg CL Loss: 0.0850
   Avg Domain Loss: 0.5360
   Source Acc: 98.81%
   Target Acc: 74.94%



Ep 16 | Cls: 0.064 | Dom Loss: 0.541 | Dom Acc: 73.6%: 100%|██████████| 211/211 [00:08<00:00, 26.01it/s]


Epoch [16/20] Stats:
   Avg CL Loss: 0.0819
   Avg Domain Loss: 0.5506
   Source Acc: 98.17%
   Target Acc: 72.16%



Ep 17 | Cls: 0.074 | Dom Loss: 0.545 | Dom Acc: 75.4%: 100%|██████████| 211/211 [00:08<00:00, 25.99it/s]


Epoch [17/20] Stats:
   Avg CL Loss: 0.0779
   Avg Domain Loss: 0.5190
   Source Acc: 98.83%
   Target Acc: 74.03%



Ep 18 | Cls: 0.057 | Dom Loss: 0.428 | Dom Acc: 87.7%: 100%|██████████| 211/211 [00:08<00:00, 26.02it/s]


Epoch [18/20] Stats:
   Avg CL Loss: 0.0694
   Avg Domain Loss: 0.5069
   Source Acc: 98.95%
   Target Acc: 80.01%



Ep 19 | Cls: 0.030 | Dom Loss: 0.508 | Dom Acc: 77.9%: 100%|██████████| 211/211 [00:08<00:00, 25.58it/s]


Epoch [19/20] Stats:
   Avg CL Loss: 0.0672
   Avg Domain Loss: 0.5067
   Source Acc: 98.80%
   Target Acc: 84.70%



Ep 20 | Cls: 0.103 | Dom Loss: 0.551 | Dom Acc: 74.4%: 100%|██████████| 211/211 [00:08<00:00, 25.80it/s]


Epoch [20/20] Stats:
   Avg CL Loss: 0.0775
   Avg Domain Loss: 0.5545
   Source Acc: 98.51%
   Target Acc: 80.21%

Run 1 Complete. Best Acc: 84.70%

  RUN 2/3 | SEED: 100
Seed set to 100
Loading pre-trained source weights from /content/drive/MyDrive/DA_Project/models/best_cnn_mnist_256_20_0.001.pth...
Loaded pre-trained source weights into DANN_CNN (features and class_classifier). Domain classifier initialized randomly.



Ep 1 | Cls: 0.035 | Dom Loss: 0.088 | Dom Acc: 100.0%: 100%|██████████| 211/211 [00:08<00:00, 24.58it/s]


Epoch [1/20] Stats:
   Avg CL Loss: 0.0160
   Avg Domain Loss: 0.3219
   Source Acc: 98.79%
   Target Acc: 49.64%



Ep 2 | Cls: 0.064 | Dom Loss: 0.102 | Dom Acc: 99.6%: 100%|██████████| 211/211 [00:08<00:00, 25.03it/s]


Epoch [2/20] Stats:
   Avg CL Loss: 0.0144
   Avg Domain Loss: 0.1235
   Source Acc: 98.84%
   Target Acc: 40.21%



Ep 3 | Cls: 0.072 | Dom Loss: 0.224 | Dom Acc: 94.3%: 100%|██████████| 211/211 [00:08<00:00, 26.04it/s]


Epoch [3/20] Stats:
   Avg CL Loss: 0.0267
   Avg Domain Loss: 0.1324
   Source Acc: 98.28%
   Target Acc: 50.43%



Ep 4 | Cls: 0.058 | Dom Loss: 0.396 | Dom Acc: 82.2%: 100%|██████████| 211/211 [00:08<00:00, 26.11it/s]


Epoch [4/20] Stats:
   Avg CL Loss: 0.0736
   Avg Domain Loss: 0.3274
   Source Acc: 98.30%
   Target Acc: 62.16%



Ep 5 | Cls: 0.030 | Dom Loss: 0.451 | Dom Acc: 80.3%: 100%|██████████| 211/211 [00:08<00:00, 25.08it/s]


Epoch [5/20] Stats:
   Avg CL Loss: 0.0831
   Avg Domain Loss: 0.4558
   Source Acc: 98.67%
   Target Acc: 67.11%



Ep 6 | Cls: 0.121 | Dom Loss: 0.606 | Dom Acc: 69.3%: 100%|██████████| 211/211 [00:08<00:00, 25.46it/s]


Epoch [6/20] Stats:
   Avg CL Loss: 0.0873
   Avg Domain Loss: 0.5046
   Source Acc: 97.70%
   Target Acc: 65.91%



Ep 7 | Cls: 0.063 | Dom Loss: 0.469 | Dom Acc: 78.3%: 100%|██████████| 211/211 [00:08<00:00, 25.52it/s]


Epoch [7/20] Stats:
   Avg CL Loss: 0.0825
   Avg Domain Loss: 0.5194
   Source Acc: 98.41%
   Target Acc: 72.08%



Ep 8 | Cls: 0.054 | Dom Loss: 0.470 | Dom Acc: 80.5%: 100%|██████████| 211/211 [00:08<00:00, 26.02it/s]


Epoch [8/20] Stats:
   Avg CL Loss: 0.0861
   Avg Domain Loss: 0.5347
   Source Acc: 98.73%
   Target Acc: 72.91%



Ep 9 | Cls: 0.067 | Dom Loss: 0.474 | Dom Acc: 80.9%: 100%|██████████| 211/211 [00:08<00:00, 26.25it/s]


Epoch [9/20] Stats:
   Avg CL Loss: 0.0769
   Avg Domain Loss: 0.4991
   Source Acc: 98.30%
   Target Acc: 73.71%



Ep 10 | Cls: 0.091 | Dom Loss: 0.506 | Dom Acc: 77.9%: 100%|██████████| 211/211 [00:08<00:00, 25.70it/s]


Epoch [10/20] Stats:
   Avg CL Loss: 0.0814
   Avg Domain Loss: 0.5223
   Source Acc: 98.66%
   Target Acc: 72.97%



Ep 11 | Cls: 0.055 | Dom Loss: 0.572 | Dom Acc: 70.3%: 100%|██████████| 211/211 [00:08<00:00, 25.15it/s]


Epoch [11/20] Stats:
   Avg CL Loss: 0.0778
   Avg Domain Loss: 0.5386
   Source Acc: 98.59%
   Target Acc: 75.91%



Ep 12 | Cls: 0.088 | Dom Loss: 0.576 | Dom Acc: 70.9%: 100%|██████████| 211/211 [00:08<00:00, 25.82it/s]


Epoch [12/20] Stats:
   Avg CL Loss: 0.0797
   Avg Domain Loss: 0.5247
   Source Acc: 98.46%
   Target Acc: 73.44%



Ep 13 | Cls: 0.108 | Dom Loss: 0.652 | Dom Acc: 63.1%: 100%|██████████| 211/211 [00:08<00:00, 25.33it/s]


Epoch [13/20] Stats:
   Avg CL Loss: 0.0835
   Avg Domain Loss: 0.5648
   Source Acc: 98.32%
   Target Acc: 71.76%



Ep 14 | Cls: 0.062 | Dom Loss: 0.498 | Dom Acc: 78.9%: 100%|██████████| 211/211 [00:08<00:00, 25.63it/s]


Epoch [14/20] Stats:
   Avg CL Loss: 0.0771
   Avg Domain Loss: 0.5636
   Source Acc: 98.80%
   Target Acc: 73.40%



Ep 15 | Cls: 0.024 | Dom Loss: 0.530 | Dom Acc: 73.8%: 100%|██████████| 211/211 [00:08<00:00, 25.81it/s]


Epoch [15/20] Stats:
   Avg CL Loss: 0.0686
   Avg Domain Loss: 0.5142
   Source Acc: 98.91%
   Target Acc: 77.91%



Ep 16 | Cls: 0.039 | Dom Loss: 0.590 | Dom Acc: 69.9%: 100%|██████████| 211/211 [00:08<00:00, 26.34it/s]


Epoch [16/20] Stats:
   Avg CL Loss: 0.0649
   Avg Domain Loss: 0.5551
   Source Acc: 98.39%
   Target Acc: 78.82%



Ep 17 | Cls: 0.086 | Dom Loss: 0.542 | Dom Acc: 73.4%: 100%|██████████| 211/211 [00:07<00:00, 26.51it/s]


Epoch [17/20] Stats:
   Avg CL Loss: 0.0782
   Avg Domain Loss: 0.5625
   Source Acc: 98.47%
   Target Acc: 77.24%



Ep 18 | Cls: 0.134 | Dom Loss: 0.518 | Dom Acc: 75.6%: 100%|██████████| 211/211 [00:08<00:00, 25.73it/s]


Epoch [18/20] Stats:
   Avg CL Loss: 0.0625
   Avg Domain Loss: 0.5191
   Source Acc: 98.69%
   Target Acc: 77.41%



Ep 19 | Cls: 0.042 | Dom Loss: 0.580 | Dom Acc: 68.6%: 100%|██████████| 211/211 [00:08<00:00, 25.68it/s]


Epoch [19/20] Stats:
   Avg CL Loss: 0.0652
   Avg Domain Loss: 0.5383
   Source Acc: 98.24%
   Target Acc: 79.51%



Ep 20 | Cls: 0.078 | Dom Loss: 0.565 | Dom Acc: 71.9%: 100%|██████████| 211/211 [00:08<00:00, 25.40it/s]


Epoch [20/20] Stats:
   Avg CL Loss: 0.0679
   Avg Domain Loss: 0.5452
   Source Acc: 98.44%
   Target Acc: 80.94%

Run 2 Complete. Best Acc: 80.94%

  RUN 3/3 | SEED: 2024
Seed set to 2024
Loading pre-trained source weights from /content/drive/MyDrive/DA_Project/models/best_cnn_mnist_256_20_0.001.pth...
Loaded pre-trained source weights into DANN_CNN (features and class_classifier). Domain classifier initialized randomly.



Ep 1 | Cls: 0.011 | Dom Loss: 0.076 | Dom Acc: 100.0%: 100%|██████████| 211/211 [00:08<00:00, 25.47it/s]


Epoch [1/20] Stats:
   Avg CL Loss: 0.0159
   Avg Domain Loss: 0.2987
   Source Acc: 99.31%
   Target Acc: 51.13%



Ep 2 | Cls: 0.003 | Dom Loss: 0.078 | Dom Acc: 100.0%: 100%|██████████| 211/211 [00:08<00:00, 25.95it/s]


Epoch [2/20] Stats:
   Avg CL Loss: 0.0165
   Avg Domain Loss: 0.1113
   Source Acc: 99.45%
   Target Acc: 33.78%



Ep 3 | Cls: 0.029 | Dom Loss: 0.146 | Dom Acc: 98.0%: 100%|██████████| 211/211 [00:07<00:00, 26.38it/s]


Epoch [3/20] Stats:
   Avg CL Loss: 0.0226
   Avg Domain Loss: 0.1132
   Source Acc: 99.23%
   Target Acc: 59.42%



Ep 4 | Cls: 0.032 | Dom Loss: 0.233 | Dom Acc: 94.3%: 100%|██████████| 211/211 [00:08<00:00, 25.47it/s]


Epoch [4/20] Stats:
   Avg CL Loss: 0.0713
   Avg Domain Loss: 0.2890
   Source Acc: 98.81%
   Target Acc: 56.98%



Ep 5 | Cls: 0.137 | Dom Loss: 0.476 | Dom Acc: 80.1%: 100%|██████████| 211/211 [00:08<00:00, 24.85it/s]


Epoch [5/20] Stats:
   Avg CL Loss: 0.0901
   Avg Domain Loss: 0.4149
   Source Acc: 98.22%
   Target Acc: 65.64%



Ep 6 | Cls: 0.088 | Dom Loss: 0.438 | Dom Acc: 83.8%: 100%|██████████| 211/211 [00:08<00:00, 24.60it/s]


Epoch [6/20] Stats:
   Avg CL Loss: 0.0994
   Avg Domain Loss: 0.4835
   Source Acc: 98.85%
   Target Acc: 78.43%



Ep 7 | Cls: 0.130 | Dom Loss: 0.530 | Dom Acc: 74.2%: 100%|██████████| 211/211 [00:08<00:00, 24.56it/s]


Epoch [7/20] Stats:
   Avg CL Loss: 0.0890
   Avg Domain Loss: 0.5038
   Source Acc: 97.74%
   Target Acc: 69.20%



Ep 8 | Cls: 0.070 | Dom Loss: 0.496 | Dom Acc: 79.3%: 100%|██████████| 211/211 [00:08<00:00, 24.95it/s]


Epoch [8/20] Stats:
   Avg CL Loss: 0.0813
   Avg Domain Loss: 0.5160
   Source Acc: 98.54%
   Target Acc: 80.66%



Ep 9 | Cls: 0.092 | Dom Loss: 0.445 | Dom Acc: 81.4%: 100%|██████████| 211/211 [00:08<00:00, 25.48it/s]


Epoch [9/20] Stats:
   Avg CL Loss: 0.0986
   Avg Domain Loss: 0.5518
   Source Acc: 98.45%
   Target Acc: 72.32%



Ep 10 | Cls: 0.051 | Dom Loss: 0.509 | Dom Acc: 77.1%: 100%|██████████| 211/211 [00:08<00:00, 25.85it/s]


Epoch [10/20] Stats:
   Avg CL Loss: 0.0917
   Avg Domain Loss: 0.5428
   Source Acc: 98.77%
   Target Acc: 74.02%



Ep 11 | Cls: 0.055 | Dom Loss: 0.450 | Dom Acc: 83.0%: 100%|██████████| 211/211 [00:08<00:00, 25.92it/s]


Epoch [11/20] Stats:
   Avg CL Loss: 0.0744
   Avg Domain Loss: 0.5023
   Source Acc: 98.75%
   Target Acc: 78.19%



Ep 12 | Cls: 0.021 | Dom Loss: 0.498 | Dom Acc: 79.1%: 100%|██████████| 211/211 [00:08<00:00, 25.75it/s]


Epoch [12/20] Stats:
   Avg CL Loss: 0.0817
   Avg Domain Loss: 0.5527
   Source Acc: 98.89%
   Target Acc: 79.63%



Ep 13 | Cls: 0.057 | Dom Loss: 0.558 | Dom Acc: 74.2%: 100%|██████████| 211/211 [00:08<00:00, 25.41it/s]


Epoch [13/20] Stats:
   Avg CL Loss: 0.0765
   Avg Domain Loss: 0.5234
   Source Acc: 98.67%
   Target Acc: 73.63%



Ep 14 | Cls: 0.105 | Dom Loss: 0.510 | Dom Acc: 76.8%: 100%|██████████| 211/211 [00:08<00:00, 25.88it/s]


Epoch [14/20] Stats:
   Avg CL Loss: 0.0801
   Avg Domain Loss: 0.5251
   Source Acc: 98.13%
   Target Acc: 74.91%



Ep 15 | Cls: 0.132 | Dom Loss: 0.490 | Dom Acc: 78.1%: 100%|██████████| 211/211 [00:08<00:00, 25.75it/s]


Epoch [15/20] Stats:
   Avg CL Loss: 0.0898
   Avg Domain Loss: 0.5474
   Source Acc: 98.26%
   Target Acc: 72.24%



Ep 16 | Cls: 0.109 | Dom Loss: 0.506 | Dom Acc: 76.8%: 100%|██████████| 211/211 [00:08<00:00, 25.64it/s]


Epoch [16/20] Stats:
   Avg CL Loss: 0.0858
   Avg Domain Loss: 0.5522
   Source Acc: 96.91%
   Target Acc: 71.49%



Ep 17 | Cls: 0.135 | Dom Loss: 0.557 | Dom Acc: 73.6%: 100%|██████████| 211/211 [00:07<00:00, 26.78it/s]


Epoch [17/20] Stats:
   Avg CL Loss: 0.0750
   Avg Domain Loss: 0.5035
   Source Acc: 98.03%
   Target Acc: 78.20%



Ep 18 | Cls: 0.042 | Dom Loss: 0.530 | Dom Acc: 74.6%: 100%|██████████| 211/211 [00:08<00:00, 26.27it/s]


Epoch [18/20] Stats:
   Avg CL Loss: 0.0767
   Avg Domain Loss: 0.5471
   Source Acc: 98.50%
   Target Acc: 71.14%



Ep 19 | Cls: 0.052 | Dom Loss: 0.535 | Dom Acc: 75.2%: 100%|██████████| 211/211 [00:08<00:00, 25.63it/s]


Epoch [19/20] Stats:
   Avg CL Loss: 0.0778
   Avg Domain Loss: 0.5444
   Source Acc: 98.34%
   Target Acc: 70.93%



Ep 20 | Cls: 0.074 | Dom Loss: 0.539 | Dom Acc: 74.2%: 100%|██████████| 211/211 [00:08<00:00, 26.00it/s]


Epoch [20/20] Stats:
   Avg CL Loss: 0.0886
   Avg Domain Loss: 0.5480
   Source Acc: 98.53%
   Target Acc: 72.73%

Run 3 Complete. Best Acc: 80.66%

 ROBUSTNESS EXPERIMENT COMPLETE WITH ANNEALING
   Individual Runs: [84.7, 80.94444444444444, 80.65555555555555]
   Final Result: 82.10% ± 1.84%


To verify the reliability of the standard annealing strategy, we repeated the experiment across three random seeds (42, 100, and 2024) using the same configuration: source pretraining at Batch 256, followed by adaptation with Alpha Annealing ($0 \to 1.0$) and Learning Rate Decay.The experiment yielded consistent results with individual accuracies of 84.70%, 80.94%, and 80.66%, resulting in a final mean accuracy of 82.10% ± 1.84%.

This result is notably lower than our 'Fixed Alpha' robustness check (89.04% ± 2.01%). The consistency of this performance gap across multiple seeds confirms that the superior performance of the Fixed Alpha strategy was not a fluke. It provides strong empirical evidence that the standard DANN approach of ramping $\alpha$ to 1.0 is suboptimal for this dataset, likely because the strong adversarial signal eventually interferes with the classification task. The 'Sufficient Alignment' provided by the weaker, fixed adversary consistently preserves more discriminative information.